{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Canvas Creation from Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import json\n",
    "import requests\n",
    "import math\n",
    "import matplotlib\n",
    "import uuid\n",
    "from textwrap import dedent\n",
    "from IPython.core.display import display, HTML\n",
    "from string import Template\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from IPython.display import display,FileLink, Markdown\n",
    "from matplotlib import colors\n",
    "from random import seed\n",
    "from random import randint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below this is arbitrary definitions for these variables, their actual values will be selected in the Appyter (so DON'T ADD BELOW CELL TO APPYTER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "bar_color = 'mediumforestgreen'\n",
    "bar_color_not_sig = 'lightgrey'\n",
    "edgecolor = None\n",
    "chart_type = \"Canvas\"\n",
    "all_libraries = ['BioCarta_2016']\n",
    "color = 'lightpink'\n",
    "\n",
    "# list of genes that was input\n",
    "genes = ['TP53', 'TNF', 'EGFR', 'GKN1', 'HADHA', 'APOE', 'ESR1', 'VEGFA', 'TGFB1', 'PREPL', 'TIA1', 'TPO', 'TTN', 'SATB2', 'CHPF', 'MALL', 'MIPIP', 'NUPL1', 'IL6', 'PDIA3', 'CTNNB1', 'SLC39A1', 'DTNA','SLC1A1', 'GALNT2', 'HIST2H2AC', 'CD63']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": [
     "outputPrepend"
    ]
   },
   "outputs": [],
   "source": [
    "def library_processing(index):\n",
    "    '''processes library data. returns anneal_list, x_dimension, y_dimension, num_hex'''\n",
    "    raw_library_data = []\n",
    "    library_data = []\n",
    "\n",
    "    library_name = all_libraries[index]\n",
    "    with open('Enrichr-Libraries/'+library_name, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            raw_library_data.append(line.split(\"\\t\\t\"))\n",
    "    name = []\n",
    "    gene_list = []\n",
    "\n",
    "    for i in range(len(raw_library_data)):\n",
    "        name += [raw_library_data[i][0]]\n",
    "        raw_genes = raw_library_data[i][1].split(\"\\t\")\n",
    "        gene_list += [raw_genes[:-1]]\n",
    "\n",
    "    library_data = [list(a) for a in zip(name, gene_list)]\n",
    "\n",
    "    # library_data: a 2D list where the first element is the name and the second element is a list      of genes associated with that name\n",
    "\n",
    "    jaccard_indices = []\n",
    "    indices = []\n",
    "\n",
    "    for gene_set in library_data:\n",
    "        intersection = [value for value in gene_set[1] if value in genes]\n",
    "        index = len(intersection)/(len(gene_set[1]) + len(genes))\n",
    "        jaccard_indices += [[gene_set[0], index]]\n",
    "        indices += [round(index, 5)]\n",
    "\n",
    "    # determine the dimensions of the canvas\n",
    "    x_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "    y_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "\n",
    "    # Zip name, gene_list, indices, and blank list for neighbor score then add dummy entries to the      zipped list\n",
    "    num_hex = x_dimension*y_dimension\n",
    "    neighbor_score = [0.0] * len(name)\n",
    "    anneal_list = list(zip(name, gene_list, indices, neighbor_score))\n",
    "\n",
    "    unzipped_anneal_list = [name, gene_list, indices, neighbor_score]\n",
    "\n",
    "    # Add \"dummy\" hexagons\n",
    "    for i in range(len(name), num_hex):\n",
    "        anneal_list += [('', [], 0.0, 0.0)]\n",
    "\n",
    "    return anneal_list, x_dimension, y_dimension, num_hex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# color options: red, blue, green. grey, orange, purple, yellow, pink\n",
    "def get_color(anneal_list):\n",
    "    r_value = 0\n",
    "    g_value = 0\n",
    "    b_value = 0\n",
    "\n",
    "    if color == 'tomato':\n",
    "        r_value = 10\n",
    "        g_value = 3\n",
    "        b_value = 3\n",
    "    if color == 'lightskyblue':\n",
    "        r_value = 0\n",
    "        g_value = 3\n",
    "        b_value = 10\n",
    "    if color == 'plum':\n",
    "        r_value = 5\n",
    "        g_value = 0\n",
    "        b_value = 10\n",
    "    if color == 'mediumspringgreen':\n",
    "        r_value = 0\n",
    "        g_value = 10\n",
    "        b_value = 2\n",
    "    if color == 'lightgrey':\n",
    "        r_value = 10\n",
    "        g_value = 10\n",
    "        b_value = 10\n",
    "    if color == 'orange':\n",
    "        r_value = 10\n",
    "        g_value = 5.5\n",
    "        b_value = 0\n",
    "    if color == 'lightpink':\n",
    "        r_value = 10\n",
    "        g_value = 0\n",
    "        b_value = 10\n",
    "    if color == 'yellow':\n",
    "        r_value = 10\n",
    "        g_value = 9\n",
    "        b_value = 0\n",
    "\n",
    "    color_list = []\n",
    "\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "\n",
    "    for i in range(x_dimension*y_dimension):\n",
    "        if i < len(unzipped_anneal_list[2]):\n",
    "            color_list += [matplotlib.colors.to_hex((unzipped_anneal_list[2][i]*r_value,\n",
    "            unzipped_anneal_list[2][i]*g_value, unzipped_anneal_list[2][i]*b_value))]\n",
    "        else:\n",
    "            color_list += [\"#000000\"]\n",
    "    return color_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Annealing  \n",
    "\n",
    "Edge cases to consider:  \n",
    "1. Corner (0, 15, 255, 240)  \n",
    "2. Edge (1-14, 16, 31, 32, etc., 241-254)   \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": [
     "outputPrepend"
    ]
   },
   "outputs": [],
   "source": [
    "def find_neighbors(ind):\n",
    "    '''returns a list of the indices of the neighbors of the index given'''\n",
    "    neighbors = []\n",
    "    if 0 <= ind <= x_dimension-1:\n",
    "        # top row (inc. corners)\n",
    "        if ind == 0:\n",
    "            # top left corner\n",
    "            neighbors = [num_hex-1, num_hex-x_dimension, x_dimension-1, 2*x_dimension-1, ind+1,                  ind+x_dimension]\n",
    "        elif ind == (x_dimension-1):\n",
    "            # top right corner\n",
    "            neighbors = [ind-1, ind+x_dimension, 0, ind+x_dimension-1, num_hex-2, num_hex-1]\n",
    "        else:\n",
    "            # non-corner top row\n",
    "            neighbors = [ind-1, ind+1, ind+x_dimension, ind+num_hex-x_dimension-1, \n",
    "            ind+num_hex-x_dimension, ind+x_dimension-1]\n",
    "\n",
    "    elif (num_hex - x_dimension) <= ind <= num_hex -1:\n",
    "        if ind == (num_hex-x_dimension):\n",
    "            # bottom left corner\n",
    "            neighbors = [ind+1, ind-x_dimension, ind-x_dimension+1, 0, 1, num_hex-1]\n",
    "        elif ind == (num_hex-1):\n",
    "            # bottom right corner\n",
    "            neighbors = [ind-1, ind-x_dimension, ind-x_dimension+1, 0, x_dimension-1,\n",
    "            num_hex-2*x_dimension]\n",
    "        else:\n",
    "            # non-corner bottom row\n",
    "            neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind-num_hex+x_dimension,\n",
    "            ind-num_hex+x_dimension+1]\n",
    "    elif ind % y_dimension == 0 and (ind/y_dimension)%2 == 1:\n",
    "        # \"inner\" left edge (not top or bottom row)\n",
    "        neighbors = [ind+x_dimension-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension+1]\n",
    "    elif ind % y_dimension == 0 and (ind/y_dimension)%2 == 0:\n",
    "        # \"outer\" left edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind+1, ind+x_dimension, ind+2*x_dimension-1, ind-x_dimension, \n",
    "        ind+x_dimension-1]\n",
    "    elif (ind+1) % y_dimension == 0 and ((ind+1)/y_dimension)%2 == 0:\n",
    "        # \"outer\" right edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind-2*x_dimension+1]\n",
    "    elif (ind+1) % y_dimension == 0 and ((ind+1)/y_dimension)%2 == 1:\n",
    "        # \"inner\" right edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind-x_dimension-1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension-1]\n",
    "    else:\n",
    "        # middle\n",
    "        neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension+1]\n",
    "    return neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initially find fitness\n",
    "def find_fitness(anneal_list):\n",
    "    fitness = 0\n",
    "    for i in range(len(anneal_list)):\n",
    "        neighbors = find_neighbors(i)\n",
    "        sum_neighbor_score = 0\n",
    "        for index in neighbors:\n",
    "            intersection = [value for value in anneal_list[index][1] if value in anneal_list[i][1]]\n",
    "            if len(anneal_list[index][1]) + len(anneal_list[i][1]) != 0:\n",
    "                jaccard = len(intersection)/(len(anneal_list[index][1]) + len(anneal_list[i][1]))\n",
    "            else:\n",
    "                jaccard = 0.0\n",
    "            sum_neighbor_score += jaccard\n",
    "        hex_list = list(anneal_list[i])\n",
    "        hex_list[3] = sum_neighbor_score\n",
    "        hex_tuple = tuple(hex_list)\n",
    "        anneal_list[i] = hex_tuple\n",
    "        fitness += sum_neighbor_score\n",
    "    return fitness, anneal_list\n",
    "\n",
    "# take indices of swapped hexagons\n",
    "def find_swapped_fitness(anneal_list, swapped_a, swapped_b, old_fitness):\n",
    "    neighbors_a = find_neighbors(swapped_a)\n",
    "    neighbors_b = find_neighbors(swapped_b)\n",
    "    hexagons_to_update = [swapped_a, swapped_b] + neighbors_a + neighbors_b\n",
    "\n",
    "    new_fitness = 0\n",
    "    # Recalculate scores for all hexagons that need updating\n",
    "    for hex in hexagons_to_update:\n",
    "\n",
    "        # subtract out the swapped neighbor fitnesses because they are changing \n",
    "        old_fitness -= anneal_list[hex][3]\n",
    "\n",
    "        neighbors = find_neighbors(hex)\n",
    "        sum_neighbor_score = 0\n",
    "        for index in neighbors:\n",
    "            intersection = [value for value in anneal_list[index][1] if value in anneal_list[hex][1]]\n",
    "            if len(anneal_list[index][1]) + len(anneal_list[hex][1]) != 0:\n",
    "                jaccard = len(intersection)/(len(anneal_list[index][1]) + len(anneal_list[hex][1]))\n",
    "            else:\n",
    "                jaccard = 0.0\n",
    "            sum_neighbor_score += jaccard\n",
    "        hex_list = list(anneal_list[hex])\n",
    "        hex_list[3] = sum_neighbor_score\n",
    "        hex_tuple = tuple(hex_list)\n",
    "        anneal_list[hex] = hex_tuple\n",
    "        new_fitness += sum_neighbor_score\n",
    "    return old_fitness + new_fitness, anneal_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unzip_list(anneal_list):\n",
    "    unzipped_list = zip(*anneal_list)\n",
    "    return list(unzipped_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annealing(anneal_list, steps, old_fitness):\n",
    "    # returns unzipped list\n",
    "    for i in range(steps):\n",
    "        index_a = randint(0, num_hex-1)\n",
    "        index_b = randint(0, num_hex-1)\n",
    "        anneal_list[index_a], anneal_list[index_b] = anneal_list[index_b], anneal_list[index_a]\n",
    "        new_fitness, anneal_list = find_swapped_fitness(anneal_list, index_a, index_b, old_fitness)\n",
    "        if new_fitness <= old_fitness:\n",
    "            # swap back\n",
    "            anneal_list[index_a], anneal_list[index_b] = anneal_list[index_b], anneal_list[index_a]\n",
    "        else:\n",
    "            # finalize the swap by resetting old_fitness\n",
    "            old_fitness = new_fitness\n",
    "    return anneal_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: your d3 code should basically be a function, in the future it would be best to take it out.\n",
    "\n",
    "def init_chart():\n",
    "  chart_id = 'mychart-' + str(uuid.uuid4())\n",
    "  display(HTML('<script src=\"/static/components/requirejs/require.js\"></script>'))\n",
    "  display(HTML(Template(dedent('''\n",
    "  <script>\n",
    "  require.config({\n",
    "    paths: {\n",
    "      'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n",
    "      'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n",
    "    },\n",
    "    shim: {\n",
    "      'd3-hexbin': ['d3']\n",
    "    }\n",
    "  })\n",
    "\n",
    "  // If we configure mychart via url, we can eliminate this define here\n",
    "  define($chart_id, ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n",
    "    return function (figure_id, numA, numB, colorList, libraryList, indices) {\n",
    "      var margin = {top: 50, right: 20, bottom: 20, left: 50},\n",
    "        width = 850 - margin.left - margin.right,\n",
    "        height = 350 - margin.top - margin.bottom;\n",
    "\n",
    "      // append the svg object to the body of the page\n",
    "      var svG = d3.select('#' + figure_id)\n",
    "        .attr(\"width\", width + margin.left + margin.right)\n",
    "        .attr(\"height\", height + margin.top + margin.bottom)\n",
    "        .append(\"g\")\n",
    "          .attr(\"transform\",\n",
    "                \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
    "      \n",
    "      //The number of columns and rows of the heatmap\n",
    "      var MapColumns = numA,\n",
    "          MapRows = numB;\n",
    "\n",
    "      //The maximum radius the hexagons can have to still fit the screen\n",
    "      var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n",
    "\n",
    "      //Calculate the center position of each hexagon\n",
    "      var points = [];\n",
    "      for (var i = 0; i < MapRows; i++) {\n",
    "          for (var j = 0; j < MapColumns; j++) {\n",
    "              var x = hexRadius * j * Math.sqrt(3)\n",
    "              //Offset each uneven row by half of a \"hex-width\" to the right\n",
    "              if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n",
    "              var y = hexRadius * i * 1.5\n",
    "              points.push([x,y])\n",
    "          }\n",
    "      }\n",
    "\n",
    "      //Set the hexagon radius\n",
    "      var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n",
    "\n",
    "      svG.append(\"g\")\n",
    "        .selectAll(\".hexagon\")\n",
    "        .data(hexbin(points))\n",
    "        .enter().append(\"path\")\n",
    "        .attr(\"class\", \"hexagon\")\n",
    "        .attr(\"d\", function (d) {\n",
    "            return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n",
    "        })\n",
    "        .attr(\"stroke\", \"white\")\n",
    "        .attr(\"stroke-width\", \"1px\")\n",
    "        .style(\"fill\", function (d,i) { return colorList[i]; })\n",
    "        .on(\"mouseover\", mover)\n",
    "        .on(\"mouseout\", mout)\n",
    "        .append(\"svg:title\")\n",
    "        .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n",
    "\n",
    "      function mover(d) {\n",
    "      d3.select(this)\n",
    "        .transition().duration(10)  \n",
    "        .style(\"fill-opacity\", 0.3)\n",
    "      }\n",
    "\n",
    "      //Mouseout function\n",
    "      function mout(d) { \n",
    "      d3.select(this)\n",
    "        .transition().duration(10)\n",
    "        .style(\"fill-opacity\", 1)\n",
    "      }\n",
    "\n",
    "    }\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({ 'chart_id': repr(chart_id) })))\n",
    "  return chart_id\n",
    "\n",
    "def Canvas(numA, numB, colorList, libraryList, indices):\n",
    "  chart_id = init_chart()\n",
    "  display(HTML(Template(dedent('''\n",
    "  <svg id=$figure_id></svg>\n",
    "  <script>\n",
    "  require([$chart_id], function(mychart) {\n",
    "    mychart($figure_id, $numA, $numB, $colorList, $libraryList, $indices)\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({\n",
    "      'chart_id': repr(chart_id),\n",
    "      'figure_id': repr('fig-' + str(uuid.uuid4())),\n",
    "      'numA': repr(numA),\n",
    "      'numB': repr(numB),\n",
    "      'colorList': repr(color_list),\n",
    "      'libraryList': repr(list(unzipped_anneal_list[0])),\n",
    "      'indices': repr(list(unzipped_anneal_list[2]))\n",
    "  })))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "<script src=\"/static/components/requirejs/require.js\"></script>"
     },
     "metadata": {}
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "\n<script>\nrequire.config({\n  paths: {\n    'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n    'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n  },\n  shim: {\n    'd3-hexbin': ['d3']\n  }\n})\n\n// If we configure mychart via url, we can eliminate this define here\ndefine('mychart-96d51282-aeec-4f14-914d-edd512a97c95', ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n  return function (figure_id, numA, numB, colorList, libraryList, indices) {\n    var margin = {top: 50, right: 20, bottom: 20, left: 50},\n      width = 850 - margin.left - margin.right,\n      height = 350 - margin.top - margin.bottom;\n\n    // append the svg object to the body of the page\n    var svG = d3.select('#' + figure_id)\n      .attr(\"width\", width + margin.left + margin.right)\n      .attr(\"height\", height + margin.top + margin.bottom)\n      .append(\"g\")\n        .attr(\"transform\",\n              \"translate(\" + margin.left + \",\" + margin.top + \")\");\n\n    //The number of columns and rows of the heatmap\n    var MapColumns = numA,\n        MapRows = numB;\n\n    //The maximum radius the hexagons can have to still fit the screen\n    var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n\n    //Calculate the center position of each hexagon\n    var points = [];\n    for (var i = 0; i < MapRows; i++) {\n        for (var j = 0; j < MapColumns; j++) {\n            var x = hexRadius * j * Math.sqrt(3)\n            //Offset each uneven row by half of a \"hex-width\" to the right\n            if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n            var y = hexRadius * i * 1.5\n            points.push([x,y])\n        }\n    }\n\n    //Set the hexagon radius\n    var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n\n    svG.append(\"g\")\n      .selectAll(\".hexagon\")\n      .data(hexbin(points))\n      .enter().append(\"path\")\n      .attr(\"class\", \"hexagon\")\n      .attr(\"d\", function (d) {\n          return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n      })\n      .attr(\"stroke\", \"white\")\n      .attr(\"stroke-width\", \"1px\")\n      .style(\"fill\", function (d,i) { return colorList[i]; })\n      .on(\"mouseover\", mover)\n      .on(\"mouseout\", mout)\n      .append(\"svg:title\")\n      .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n\n    function mover(d) {\n    d3.select(this)\n      .transition().duration(10)  \n      .style(\"fill-opacity\", 0.3)\n    }\n\n    //Mouseout function\n    function mout(d) { \n    d3.select(this)\n      .transition().duration(10)\n      .style(\"fill-opacity\", 1)\n    }\n\n  }\n})\n</script>\n"
     },
     "metadata": {}
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "\n<svg id='fig-487b3308-4050-4e36-b1ae-e3a7c61b4ba3'></svg>\n<script>\nrequire(['mychart-96d51282-aeec-4f14-914d-edd512a97c95'], function(mychart) {\n  mychart('fig-487b3308-4050-4e36-b1ae-e3a7c61b4ba3', 16, 16, ['#350035', '#000000', '#000000', '#3e003e', '#000000', '#3a003a', '#200020', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#400040', '#000000', '#000000', '#470047', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#220022', '#000000', '#000000', '#3e003e', '#000000', '#000000', '#2b002b', '#2a002a', '#000000', '#240024', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#410041', '#250025', '#000000', '#000000', '#320032', '#270027', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#330033', '#000000', '#000000', '#000000', '#000000', '#000000', '#2f002f', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#470047', '#000000', '#000000', '#800080', '#000000', '#000000', '#000000', '#000000', '#390039', '#000000', '#000000', '#000000', '#000000', '#600060', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#490049', '#000000', '#000000', '#000000', '#000000', '#000000', '#400040', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#500050', '#340034', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#490049', '#8b008b', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#2d002d', '#000000', '#000000', '#000000', '#000000', '#000000', '#340034', '#000000', '#000000', '#360036', '#000000', '#000000', '#200020', '#000000', '#3d003d', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#400040', '#000000', '#000000', '#000000', '#400040', '#3e003e', '#000000', '#000000', '#000000', '#000000', '#330033', '#790079', '#000000', '#000000', '#000000', '#000000', '#470047', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#430043', '#790079', '#000000', '#000000', '#000000', '#000000', '#000000', '#5b005b', '#2b002b', '#000000', '#000000', '#000000', '#360036', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#360036', '#000000', '#000000', '#000000', '#000000', '#000000', '#3a003a', '#2d002d', '#000000', '#000000', '#000000', '#000000', '#000000', '#8a008a', '#000000', '#000000', '#360036', '#000000', '#000000', '#000000', '#000000', '#000000', '#350035', '#510051', '#000000', '#410041', '#000000', '#3b003b', '#000000', '#000000', '#000000', '#2d002d', '#000000', '#000000', '#3b003b', '#000000', '#330033', '#000000', '#000000', '#390039', '#390039', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#2f002f', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000'], ['NF-kB Signaling Pathway Homo sapiens h nfkbPathway', \"Hemoglobin's Chaperone Homo sapiens h ahspPathway\", 'Proteolysis and Signaling Pathway of Notch  Homo sapiens h notchpathway', 'Double Stranded RNA Induced Gene Expression Homo sapiens h rnaPathway', 'Acetylation and Deacetylation of RelA in Nucleus Homo sapiens h RELAPathway', 'TNFR1 Signaling Pathway Homo sapiens h tnfr1Pathway', 'HIV-1 Nef: negative effector of Fas and TNF Homo sapiens h HivnefPathway', 'Role of MAL in Rho-Mediated Activation of SRF Homo sapiens h malPathway', 'Role of Ran in mitotic spindle regulation Homo sapiens h ranMSpathway', 'Links between Pyk2 and Map Kinases Homo sapiens h pyk2Pathway', 'CCR3 signaling in Eosinophils Homo sapiens h CCR3Pathway', 'RNA polymerase III transcription Homo sapiens h RNApol3Pathway', 'Ras Signaling Pathway Homo sapiens h rasPathway', 'fMLP induced chemokine gene expression in HMC-1 cells Homo sapiens h fMLPpathway', 'Keratinocyte Differentiation Homo sapiens h keratinocytePathway', 'Neuroregulin receptor degredation protein-1 Controls ErbB3 receptor recycling Homo sapiens h ErbB3Pathway', 'Sumoylation by RanBP2 Regulates Transcriptional Repression Homo sapiens h ranbp2Pathway', 'West Nile Virus Homo sapiens h WNVpathway', 'uCalpain and friends in Cell spread Homo sapiens h ucalpainPathway', 'CD40L Signaling Pathway Homo sapiens h cd40Pathway', 'Role of PPAR-gamma Coactivators in Obesity and Thermogenesis Homo sapiens h ppargPathway', 'Transcription factor CREB and its extracellular signals Homo sapiens h crebPathway', 'Reversal of Insulin Resistance by Leptin Homo sapiens h leptinPathway', 'Mechanism of Protein Import into the Nucleus Homo sapiens h npcPathway', 'Agrin in Postsynaptic Differentiation Homo sapiens h agrPathway', 'Nuclear receptors coordinate the activities of chromatin remodeling complexes and coactivators to facilitate initiation of transcription in carcinoma cells Homo sapiens h rarrxrPathway', 'Sonic Hedgehog Receptor Ptc1 Regulates cell cycle Homo sapiens h ptc1Pathway', 'Apoptotic Signaling in Response to DNA Damage Homo sapiens h chemicalPathway', 'EPO Signaling Pathway Homo sapiens h epoPathway', 'MAPKinase Signaling Pathway Homo sapiens h mapkPathway', 'Role of EGF Receptor Transactivation by GPCRs in Cardiac Hypertrophy Homo sapiens h cardiacegfPathway', 'Angiotensin II mediated activation of JNK Pathway via Pyk2 dependent signaling Homo sapiens h At1rPathway', 'TPO Signaling Pathway Homo sapiens h TPOPathway', 'NFAT and Hypertrophy of the heart Homo sapiens h nfatPathway', '', 'Rho cell motility signaling pathway Homo sapiens h rhoPathway', 'Deregulation of CDK5 in Alzheimers Disease Homo sapiens h p35alzheimersPathway', 'Eukaryotic protein translation Homo sapiens h eifPathway', 'Apoptotic DNA fragmentation and tissue homeostasis Homo sapiens h DNAfragmentPathway', 'Spliceosomal Assembly Homo sapiens h smPathway', 'Overview of telomerase RNA component gene hTerc Transcriptional Regulation Homo sapiens h tercPathway', 'Thrombin signaling and protease-activated receptors Homo sapiens h Par1Pathway', 'CARM1 and Regulation of the Estrogen Receptor Homo sapiens h carm-erPathway', 'Actions of Nitric Oxide in the Heart Homo sapiens h no1Pathway', '', 'Phospholipids as signalling intermediaries Homo sapiens h edg1Pathway', 'Segmentation Clock Homo sapiens h hesPathway', 'Inactivation of Gsk3 by AKT causes accumulation of b-catenin in Alveolar Macrophages Homo sapiens h gsk3Pathway', 'The Co-Stimulatory Signal During T-cell Activation Homo sapiens h ctla4Pathway', 'Role of Mitochondria in Apoptotic Signaling Homo sapiens h mitochondriaPathway', 'Intrinsic Prothrombin Activation Pathway Homo sapiens h intrinsicPathway', '', 'Lectin Induced Complement Pathway Homo sapiens h lectinPathway', 'How does salmonella hijack a cell Homo sapiens h salmonellaPathway', 'Tumor Suppressor Arf Inhibits Ribosomal Biogenesis Homo sapiens h arfPathway', '', 'Basic Mechanisms of SUMOylation Homo sapiens h sumoPathway', 'Erythropoietin mediated neuroprotection through NF-kB Homo sapiens h eponfkbPathway', 'Rho-Selective Guanine Exchange Factor AKAP13 Mediates Stress Fiber Formation Homo sapiens h akap13Pathway', 'Corticosteroids and cardioprotection Homo sapiens h gcrPathway', 'ALK in cardiac myocytes Homo sapiens h alkPathway', 'Oxidative Stress Induced Gene Expression Via Nrf2 Homo sapiens h arenrf2Pathway', 'Stathmin and breast cancer resistance to antimicrotubule agents Homo sapiens h stathminPathway', 'PDGF Signaling Pathway Homo sapiens h pdgfPathway', 'Cyclins and Cell Cycle Regulation Homo sapiens h cellcyclePathway', 'Prion Pathway Homo sapiens h prionPathway', 'Beta-arrestins in GPCR Desensitization Homo sapiens h bArrestinPathway', 'BTG family proteins and cell cycle regulation Homo sapiens h btg2Pathway', 'Cystic fibrosis transmembrane conductance regulator (CFTR) and beta 2 adrenergic receptor (b2AR) pathway Homo sapiens h cftrPathway', 'Inhibition of Matrix Metalloproteinases Homo sapiens h reckPathway', 'IL-10 Anti-inflammatory Signaling Pathway Homo sapiens h il10Pathway', 'Phosphoinositides and their downstream targets. Homo sapiens h ptdinsPathway', 'Eicosanoid Metabolism Homo sapiens h eicosanoidPathway', 'Regulation of PGC-1a Homo sapiens h pgc1aPathway', 'Y branching of actin filaments Homo sapiens h actinYPathway', 'Chaperones modulate interferon Signaling Pathway Homo sapiens h tidPathway', '', 'Protein Kinase A at the Centrosome Homo sapiens h akapCentrosomePathway', 'AKAP95 role in mitosis and chromosome dynamics Homo sapiens h akap95Pathway', 'IL 4 signaling pathway Homo sapiens h il4Pathway', 'Cell Cycle: G1/S Check Point  Homo sapiens h g1Pathway', 'Internal Ribosome entry pathway Homo sapiens h iresPathway', '', 'mTOR Signaling Pathway Homo sapiens h mTORPathway', 'Transcription Regulation by Methyltransferase of CARM1 Homo sapiens h carm1Pathway', 'Granzyme A mediated Apoptosis Pathway Homo sapiens h setPathway', 'Regulation of Splicing through Sam68 Homo sapiens h sam68Pathway', '', 'Lissencephaly gene (LIS1) in neuronal migration and development Homo sapiens h Lis1Pathway', 'CDK Regulation of DNA Replication Homo sapiens h mcmPathway', 'Visceral Fat Deposits and the Metabolic Syndrome Homo sapiens h vobesityPathway', 'BRCA1-dependent Ub-ligase activity Homo sapiens h bard1Pathway', 'NO2-dependent IL 12 Pathway in NK cells Homo sapiens h no2il12Pathway', 'IL22 Soluble Receptor Signaling Pathway  Homo sapiens h il22bppathway', '', '', 'p53 Signaling Pathway Homo sapiens h p53Pathway', 'E2F1 Destruction Pathway Homo sapiens h skp2e2fPathway', 'Lck and Fyn tyrosine kinases in initiation of TCR Activation Homo sapiens h tcraPathway', 'Metabolism of Anandamide, an Endogenous Cannabinoid Homo sapiens h cb1rPathway', 'TNFR2 Signaling Pathway Homo sapiens h tnfr2Pathway', 'IFN alpha signaling pathway Homo sapiens h ifnaPathway', 'Influence of Ras and Rho proteins on G1 to S Transition Homo sapiens h RacCycDPathway', 'Activation of PKC through G-protein coupled receptors Homo sapiens h pkcPathway', 'B Cell Survival Pathway Homo sapiens h bcellsurvivalPathway', 'Classical Complement Pathway Homo sapiens h classicPathway', 'Regulation of eIF2 Homo sapiens h eif2Pathway', 'Pelp1 Modulation of Estrogen Receptor Activity Homo sapiens h pelp1Pathway', 'Cell Cycle: G2/M Checkpoint Homo sapiens h g2Pathway', 'How Progesterone Initiates the Oocyte Maturation Homo sapiens h mPRPathway', 'PKC-catalyzed phosphorylation of inhibitory phosphoprotein of myosin phosphatase Homo sapiens h myosinPathway', 'Antigen Processing and Presentation Homo sapiens h mhcPathway', 'Melanocyte Development and Pigmentation Pathway Homo sapiens h melanocytepathway', 'Trka Receptor Signaling Pathway Homo sapiens h trkaPathway', 'PTEN dependent cell cycle arrest and apoptosis Homo sapiens h ptenPathway', 'The IGF-1 Receptor and Longevity Homo sapiens h longevityPathway', 'CBL mediated ligand-induced downregulation of EGF receptors Homo sapiens h cblPathway', 'Role of ERBB2 in Signal Transduction and Oncology Homo sapiens h her2Pathway', 'Role of PI3K subunit p85 in regulation of Actin Organization and Cell Migration Homo sapiens h cdc42racPathway', 'ChREBP regulation by carbohydrates and cAMP Homo sapiens h chrebpPathway', 'Gamma-aminobutyric Acid Receptor Life Cycle Homo sapiens h gabaPathway', 'IFN gamma signaling pathway  Homo sapiens h ifngPathway', '', 'Erk and PI-3 Kinase Are Necessary for Collagen Binding in Corneal Epithelia Homo sapiens h ecmPathway', 'mCalpain and friends in Cell motility Homo sapiens h mCalpainPathway', 'Signaling Pathway from G-Protein Families Homo sapiens h gpcrPathway', 'Ras-Independent pathway in NK cell-mediated cytotoxicity Homo sapiens h nkcellsPathway', 'Role of Parkin in Ubiquitin-Proteasomal Pathway Homo sapiens h parkinPathway', 'SUMOylation as a mechanism to modulate CtBP-dependent gene responses Homo sapiens h ctbp1Pathway', 'ER?associated degradation (ERAD) Pathway Homo sapiens h eradPathway', 'Role of BRCA1,  BRCA2 and ATR in Cancer Susceptibility Homo sapiens h atrbrcaPathway', 'Extrinsic Prothrombin Activation Pathway Homo sapiens h extrinsicPathway', 'Integrin Signaling Pathway Homo sapiens h integrinPathway', 'EGF Signaling Pathway Homo sapiens h egfPathway', 'Role of MEF2D in T-cell Apoptosis Homo sapiens h mef2dPathway', 'IL-7 Signal Transduction Homo sapiens h il7Pathway', 'Mechanism of Gene Regulation by Peroxisome Proliferators via PPARa Homo sapiens h pparaPathway', 'TACI and BCMA stimulation of B cell immune responses Homo sapiens h tall1Pathway', 'Telomeres, Telomerase, Cellular Aging, and  Immortality Homo sapiens h telPathway', 'Human Cytomegalovirus and Map Kinase Pathways Homo sapiens h hcmvPathway', 'Control of Gene Expression by Vitamin D Receptor Homo sapiens h vdrPathway', 'BCR Signaling Pathway Homo sapiens h bcrPathway', 'Phospholipase C-epsilon pathway Homo sapiens h plcePathway', 'Erk1/Erk2 Mapk Signaling pathway Homo sapiens h erkPathway', 'IL 3 signaling pathway Homo sapiens h il3Pathway', 'IL 6 signaling pathway Homo sapiens h il6Pathway', 'Skeletal muscle hypertrophy is regulated via AKT/mTOR pathway Homo sapiens h igf1mtorpathway', 'The PRC2 Complex Sets Long-term Gene Silencing Through Modification of Histone Tails Homo sapiens h prc2Pathway', 'Ca++/ Calmodulin-dependent Protein Kinase Activation Homo sapiens h CaCaMPathway', 'RB Tumor Suppressor/Checkpoint Signaling in response to DNA damage Homo sapiens h rbPathway', 'Stress Induction of HSP Regulation Homo sapiens h hsp27Pathway', 'METS affect on Macrophage Differentiation Homo sapiens h etsPathway', 'Regulation of MAP Kinase Pathways Through Dual Specificity Phosphatases Homo sapiens h dspPathway', 'Fc Epsilon Receptor I Signaling in Mast Cells Homo sapiens h fcer1Pathway', 'Calcium Signaling by HBx of Hepatitis B virus Homo sapiens h HBxPathway', 'TNF/Stress Related Signaling Homo sapiens h stressPathway', 'Signal transduction through IL1R Homo sapiens h il1rPathway', 'Platelet Amyloid Precursor Protein Pathway Homo sapiens h plateletAppPathway', 'Roles of Beta-arrestin-dependent Recruitment of Src Kinases in GPCR Signaling Homo sapiens h bArrestin-srcPathway', 'Pertussis toxin-insensitive CCR5 Signaling in Macrophage Homo sapiens h Ccr5Pathway', 'Inhibition of Cellular Proliferation by Gleevec Homo sapiens h gleevecpathway', 'SODD/TNFR1 Signaling Pathway Homo sapiens h soddPathway', 'Role of Beta-arrestins in the activation and targeting of MAP kinases Homo sapiens h barr-mapkPathway', 'p38 MAPK Signaling Pathway  Homo sapiens h p38mapkPathway', 'Nerve growth factor pathway (NGF) Homo sapiens h ngfPathway', 'AKT Signaling Pathway Homo sapiens h aktPathway', 'Caspase Cascade in Apoptosis Homo sapiens h caspasePathway', 'cdc25 and chk1 Regulatory Pathway in response to DNA damage Homo sapiens h cdc25Pathway', '', 'Cycling of Ran in nucleocytoplasmic transport Homo sapiens h ranPathway', 'Toll-Like Receptor Pathway Homo sapiens h tollPathway', 'Regulation of transcriptional activity by PML Homo sapiens h pmlPathway', 'Estrogen-responsive protein Efp controls cell cycle and breast tumors growth Homo sapiens h EfpPathway', 'Aspirin Blocks Signaling Pathway Involved in Platelet Activation Homo sapiens h sppaPathway', 'G-Protein Signaling Through Tubby Proteins Homo sapiens h tubbyPathway', 'ADP-Ribosylation Factor Homo sapiens h arapPathway', 'Ion Channels and Their Functional Role in Vascular Endothelium Homo sapiens h raccPathway', 'IL 2 signaling pathway Homo sapiens h il2Pathway', 'NFkB activation by Nontypeable Hemophilus influenzae Homo sapiens h nthiPathway', 'Ceramide Signaling Pathway Homo sapiens h ceramidePathway', 'The information-processing pathway at the IFN-beta enhancer Homo sapiens h pcafpathway', '', 'Wnt/LRP6 Signalling Homo sapiens h wnt-lrp6Pathway', 'Downregulated of MTA-3 in ER-negative Breast Tumors Homo sapiens h mta3Pathway', 'Growth Hormone Signaling Pathway Homo sapiens h ghPathway', 'SREBP control of lipid synthesis Homo sapiens h s1pPathway', 'FAS signaling pathway ( CD95 ) Homo sapiens h fasPathway', '', 'Basic mechanism of action of PPARa, PPARb(d) and PPARg and effects on gene expression Homo sapiens h pparPathway', 'Polyadenylation of mRNA Homo sapiens h cpsfPathway', 'Nitric Oxide Signaling Pathway Homo sapiens h nos1Pathway', 'Activation of Csk by cAMP-dependent Protein Kinase Inhibits Signaling through the T Cell Receptor Homo sapiens h CSKPathway', 'Mechanism of Acetaminophen Activity and Toxicity Homo sapiens h AcetaminophenPathway', 'Regulation of eIF4e and p70 S6 Kinase Homo sapiens h eif4Pathway', 'Stat3 Signaling Pathway Homo sapiens h stat3Pathway', 'The 4-1BB-dependent immune response Homo sapiens h 41bbPathway', 'Fibrinolysis Pathway Homo sapiens h fibrinolysisPathway', 'Signal Dependent Regulation of Myogenesis by Corepressor MITR Homo sapiens h MITRPathway', 'Sprouty regulation of tyrosine kinase signals Homo sapiens h spryPathway', 'Bioactive Peptide Induced Signaling Pathway Homo sapiens h biopeptidesPathway', 'Insulin Signaling Pathway Homo sapiens h insulinPathway', 'Multiple antiapoptotic pathways from IGF-1R signaling lead to BAD phosphorylation Homo sapiens h igf1rPathway', 'Cyclin E Destruction Pathway Homo sapiens h fbw7Pathway', 'Phospholipase C Signaling Pathway Homo sapiens h plcPathway', 'Map Kinase Inactivation of SMRT Corepressor Homo sapiens h egfr smrtePathway', 'VEGF, Hypoxia, and Angiogenesis Homo sapiens h vegfPathway', 'Induction of apoptosis through DR3 and DR4/5 Death Receptors  Homo sapiens h deathPathway', 'FOSB gene expression and drug abuse Homo sapiens h fosbPathway', 'Cardiac Protection Against ROS  Homo sapiens h flumazenilPathway', 'Rac 1 cell motility signaling pathway Homo sapiens h rac1Pathway', 'D4-GDI Signaling Pathway Homo sapiens h d4gdiPathway', 'Overview of telomerase protein component gene hTert Transcriptional Regulation  Homo sapiens h tertpathway', 'Multi-Drug Resistance Factors Homo sapiens h mrpPathway', '', 'TGF beta signaling pathway Homo sapiens h tgfbPathway', 'Role of Erk5 in Neuronal Survival Homo sapiens h erk5Pathway', '', 'T Cell Receptor Signaling Pathway Homo sapiens h tcrPathway', 'Role of nicotinic acetylcholine receptors in the regulation of apoptosis Homo sapiens h achPathway', 'CXCR4 Signaling Pathway Homo sapiens h cxcr4Pathway', 'Hypoxia and p53 in the Cardiovascular system Homo sapiens h p53hypoxiaPathway', 'Trefoil Factors Initiate  Mucosal Healing Homo sapiens h tffPathway', 'IL12 and Stat4 Dependent Signaling Pathway in Th1 Development Homo sapiens h IL12Pathway', 'Cell to Cell Adhesion Signaling Homo sapiens h cell2cellPathway', 'GATA3 participate in activating the Th2 cytokine genes expression Homo sapiens h GATA3pathway', 'Hypoxia-Inducible Factor in the Cardiovascular System Homo sapiens h hifPathway', 'Activation of cAMP-dependent protein kinase, PKA Homo sapiens h gsPathway', 'Regulation of p27 Phosphorylation during Cell Cycle Progression Homo sapiens h p27Pathway', 'Visual Signal Transduction Homo sapiens h rhodopsinPathway', 'WNT Signaling Pathway Homo sapiens h wntPathway', '', 'IL-2 Receptor Beta Chain in T cell Activation Homo sapiens h il2rbPathway', 'Cadmium induces DNA synthesis and proliferation in macrophages Homo sapiens h cdMacPathway', 'Control of skeletal myogenesis by HDAC and calcium/calmodulin-dependent kinase (CaMK) Homo sapiens h hdacPathway', 'CTCF: First Multivalent Nuclear Factor Homo sapiens h ctcfPathway', '', 'Endocytotic role of NDK, Phosphins and Dynamin Homo sapiens h ndkDynaminPathway', 'Regulation of cell cycle progression by Plk3 Homo sapiens h plk3Pathway', 'ATM Signaling Pathway Homo sapiens h atmPathway', 'Regulators of Bone Mineralization Homo sapiens h npp1Pathway', 'Regulation of BAD phosphorylation Homo sapiens h badPathway', 'g-Secretase mediated ErbB4 Signaling Pathway  Homo sapiens h erbB4pathway', 'Regulation of ck1/cdk5 by type 1 glutamate receptors Homo sapiens h ck1Pathway', 'Effects of calcineurin in Keratinocyte Differentiation Homo sapiens h calcineurinPathway', 'Bone Remodeling Homo sapiens h ranklPathway', 'Multi-step Regulation of Transcription by Pitx2 Homo sapiens h pitx2Pathway', 'Chromatin Remodeling by hSWI/SNF ATP-dependent Complexes Homo sapiens h hSWI-SNFpathway', '', 'Phosphorylation of MEK1 by cdk5/p35 down regulates the MAP kinase pathway Homo sapiens h cdk5Pathway', 'IGF-1 Signaling Pathway Homo sapiens h igf1Pathway', 'Attenuation of GPCR Signaling Homo sapiens h agpcrPathway', 'Regulation of Spermatogenesis by CREM Homo sapiens h cremPathway', 'Alternative Complement Pathway Homo sapiens h alternativePathway', '', 'Repression of Pain Sensation by the Transcriptional Regulator DREAM Homo sapiens h dreampathway', 'TSP-1 Induced Apoptosis in Microvascular Endothelial Cell  Homo sapiens h tsp1Pathway'], [0.02083, 0.0, 0.0, 0.02439, 0.0, 0.02273, 0.0125, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.025, 0.0, 0.0, 0.02778, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01351, 0.0, 0.0, 0.02439, 0.0, 0.0, 0.01667, 0.01639, 0.0, 0.01408, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02564, 0.01449, 0.0, 0.0, 0.01961, 0.01515, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01852, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02778, 0.0, 0.0, 0.05, 0.0, 0.0, 0.0, 0.0, 0.02222, 0.0, 0.0, 0.0, 0.0, 0.03774, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02857, 0.0, 0.0, 0.0, 0.0, 0.0, 0.025, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.03125, 0.02041, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02857, 0.05455, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01754, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02041, 0.0, 0.0, 0.02128, 0.0, 0.0, 0.01266, 0.0, 0.02381, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.025, 0.0, 0.0, 0.0, 0.025, 0.02439, 0.0, 0.0, 0.0, 0.0, 0.02, 0.04762, 0.0, 0.0, 0.0, 0.0, 0.02778, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02632, 0.04762, 0.0, 0.0, 0.0, 0.0, 0.0, 0.03571, 0.01667, 0.0, 0.0, 0.0, 0.02128, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02128, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02273, 0.01754, 0.0, 0.0, 0.0, 0.0, 0.0, 0.05405, 0.0, 0.0, 0.02128, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02083, 0.03175, 0.0, 0.02564, 0.0, 0.02326, 0.0, 0.0, 0.0, 0.01754, 0.0, 0.0, 0.02326, 0.0, 0.02, 0.0, 0.0, 0.02222, 0.02222, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01852, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])\n})\n</script>\n"
     },
     "metadata": {}
    }
   ],
   "source": [
    "anneal_list, x_dimension, y_dimension, num_hex = library_processing(0)\n",
    "anneal_list = annealing(anneal_list, 5000, find_fitness(anneal_list)[0])\n",
    "color_list = get_color(anneal_list)\n",
    "unzipped_anneal_list = unzip_list(anneal_list)\n",
    "Canvas(x_dimension, y_dimension, color_list, list(unzipped_anneal_list[0]), list(unzipped_anneal_list[2]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('base': conda)",
   "language": "python",
   "name": "python37664bitbasecondad81f4d096e8b4b9885e3d702e0bb3be1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}