{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Canvas Creation from Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import json\n",
    "import requests\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import time\n",
    "import math\n",
    "import matplotlib\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from IPython.display import display,FileLink, Markdown\n",
    "from matplotlib import colors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below this is arbitrary definitions for these variables, their actual values will be selected in the Appyter (so DON'T ADD BELOW CELL TO APPYTER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "bar_color = 'mediumforestgreen'\n",
    "bar_color_not_sig = 'lightgrey'\n",
    "edgecolor = None\n",
    "chart_type = \"Canvas\"\n",
    "all_libraries = ['BioCarta_2015']\n",
    "color = 'lightpink'\n",
    "\n",
    "# list of genes that was input\n",
    "genes = ['TP53', 'TNF', 'EGFR', 'GKN1', 'HADHA', 'APOE', 'ESR1', 'VEGFA', 'TGFB1', 'PREPL', 'TIA1', 'TPO', 'TTN', 'SATB2', 'CHPF', 'MALL', 'MIPIP', 'NUPL1', 'IL6', 'PDIA3', 'CTNNB1', 'SLC39A1', 'DTNA','SLC1A1', 'GALNT2', 'HIST2H2AC', 'CD63']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "tags": [
     "outputPrepend"
    ]
   },
   "outputs": [],
   "source": [
    "# NOTE: currently only covers case where these is one library\n",
    "raw_library_data = []\n",
    "library_data = []\n",
    "\n",
    "for library_name in all_libraries:\n",
    "    with open('Enrichr-Libraries/'+library_name, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            raw_library_data.append(line.split(\"\\t\\t\"))\n",
    "name = []\n",
    "gene_list = []\n",
    "\n",
    "for i in range(len(raw_library_data)):\n",
    "    name += [raw_library_data[i][0]]\n",
    "    raw_genes = raw_library_data[i][1].split(\"\\t\")\n",
    "    gene_list += [raw_genes[:-1]]\n",
    "\n",
    "library_data = [list(a) for a in zip(name, gene_list)]\n",
    "\n",
    "# library_data: a 2D list where the first element is the name and the second element is a list of genes associated with that name\n",
    "\n",
    "jaccard_indices = []\n",
    "indices = []\n",
    "\n",
    "for gene_set in library_data:\n",
    "    intersection = [value for value in gene_set[1] if value in genes]\n",
    "    index = len(intersection)/(len(gene_set[1]) + len(genes))\n",
    "    jaccard_indices += [[gene_set[0], index]]\n",
    "    indices += [round(index, 5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "# determine the dimensions of the canvas\n",
    "# currently only makes squares which is not the most efficient\n",
    "\n",
    "x_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "y_dimension = math.ceil(math.sqrt(len(indices)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "# color options: red, blue, green. grey, orange, purple, yellow, pink\n",
    "r_value = 0\n",
    "g_value = 0\n",
    "b_value = 0\n",
    "\n",
    "if color == 'tomato':\n",
    "    r_value = 10\n",
    "    g_value = 3\n",
    "    b_value = 3\n",
    "if color == 'lightskyblue':\n",
    "    r_value = 0\n",
    "    g_value = 3\n",
    "    b_value = 10\n",
    "if color == 'plum':\n",
    "    r_value = 5\n",
    "    g_value = 0\n",
    "    b_value = 10\n",
    "if color == 'mediumspringgreen':\n",
    "    r_value = 0\n",
    "    g_value = 10\n",
    "    b_value = 2\n",
    "if color == 'lightgrey':\n",
    "    r_value = 10\n",
    "    g_value = 10\n",
    "    b_value = 10\n",
    "if color == 'orange':\n",
    "    r_value = 10\n",
    "    g_value = 5.5\n",
    "    b_value = 0\n",
    "if color == 'lightpink':\n",
    "    r_value = 10\n",
    "    g_value = 0\n",
    "    b_value = 10\n",
    "if color == 'yellow':\n",
    "    r_value = 10\n",
    "    g_value = 9\n",
    "    b_value = 0\n",
    "\n",
    "color_list = []\n",
    "\n",
    "for i in range(x_dimension*y_dimension):\n",
    "    if i < len(indices):\n",
    "        color_list += [matplotlib.colors.to_hex((indices[i]*r_value, indices[i]*g_value, \n",
    "        indices[i]*b_value))]\n",
    "    else:\n",
    "        color_list += [\"#000000\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Annealing  \n",
    "\n",
    "Edge cases to consider:  \n",
    "1. Corner (0, 15, 255, 240)  \n",
    "2. Edge (1-14, 16, 31, 32, etc., 241-254)   \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Zip name, gene_list, indices, and blank list for neighbor score then add dummy entries to the zipped list\n",
    "num_hex = x_dimension*y_dimension\n",
    "neighbor_score = [0.0] * num_hex\n",
    "anneal_list = list(zip(name, gene_list, indices, neighbor_score))\n",
    "\n",
    "def find_neighbors(ind):\n",
    "    '''returns a list of the indices of the neighbors of the index given'''\n",
    "    neighbors = []\n",
    "    if 0 <= ind <= x_dimension-1:\n",
    "        # top row (inc. corners)\n",
    "        if ind == 0:\n",
    "            # top left corner\n",
    "            neighbors = [num_hex-1, num_hex-x_dimension, x_dimension-1, 2*x_dimension-1, ind+1,                  ind+x_dimension]\n",
    "        elif ind == (x_dimension-1):\n",
    "            # top right corner\n",
    "            neighbors = [ind-1, ind+x_dimension, 0, ind+x_dimension-1, num_hex-2, num_hex-1]\n",
    "        else:\n",
    "            # non-corner top row\n",
    "            neighbors = [ind-1, ind+1, ind+x_dimension, ind+num_hex-x_dimension-1, \n",
    "            ind+num_hex-x_dimension, ind+x_dimension-1]\n",
    "\n",
    "    elif (num_hex - x_dimension) <= ind <= num_hex -1:\n",
    "        if ind == (num_hex-x_dimension):\n",
    "            # bottom left corner\n",
    "            neighbors = [ind+1, ind-x_dimension, ind-x_dimension+1, 0, 1, num_hex-1]\n",
    "        elif ind == (num_hex-1):\n",
    "            # bottom right corner\n",
    "            neighbors = [ind-1, ind-x_dimension, ind-x_dimension+1, 0, x_dimension-1,\n",
    "            num_hex-2*x_dimension]\n",
    "        else:\n",
    "            # non-corner bottom row\n",
    "            neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind-num_hex+x_dimension,\n",
    "            ind-num_hex+x_dimension+1]\n",
    "    elif ind % y_dimension == 0 and (ind/y_dimension)%2 == 1:\n",
    "        # \"inner\" left edge (not top or bottom row)\n",
    "        neighbors = [ind+x_dimension-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension+1]\n",
    "    elif ind % y_dimension == 0 and (ind/y_dimension)%2 == 0:\n",
    "        # \"outer\" left edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind+1, ind+x_dimension, ind+2*x_dimension-1, ind-x_dimension, \n",
    "        ind+x_dimension-1]\n",
    "    elif (ind+1) % y_dimension == 0 and ((ind+1)/y_dimension)%2 == 0:\n",
    "        # \"outer\" right edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind-2*x_dimension+1]\n",
    "    elif (ind+1) % y_dimension == 0 and ((ind+1)/y_dimension)%2 == 1:\n",
    "        # \"inner\" right edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind-x_dimension-1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension-1]\n",
    "    else:\n",
    "        # middle\n",
    "        neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension+1]\n",
    "    return neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "import uuid\n",
    "from textwrap import dedent\n",
    "from IPython.core.display import display, HTML\n",
    "from string import Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: your d3 code should basically be a function, in the future it would be best to take it out.\n",
    "\n",
    "def init_chart():\n",
    "  chart_id = 'mychart-' + str(uuid.uuid4())\n",
    "  display(HTML('<script src=\"/static/components/requirejs/require.js\"></script>'))\n",
    "  display(HTML(Template(dedent('''\n",
    "  <script>\n",
    "  require.config({\n",
    "    paths: {\n",
    "      'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n",
    "      'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n",
    "    },\n",
    "    shim: {\n",
    "      'd3-hexbin': ['d3']\n",
    "    }\n",
    "  })\n",
    "\n",
    "  // If we configure mychart via url, we can eliminate this define here\n",
    "  define($chart_id, ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n",
    "    return function (figure_id, numA, numB, colorList, libraryList, indices) {\n",
    "      var margin = {top: 50, right: 20, bottom: 20, left: 50},\n",
    "        width = 850 - margin.left - margin.right,\n",
    "        height = 350 - margin.top - margin.bottom;\n",
    "\n",
    "      // append the svg object to the body of the page\n",
    "      var svG = d3.select('#' + figure_id)\n",
    "        .attr(\"width\", width + margin.left + margin.right)\n",
    "        .attr(\"height\", height + margin.top + margin.bottom)\n",
    "        .append(\"g\")\n",
    "          .attr(\"transform\",\n",
    "                \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
    "      \n",
    "      //The number of columns and rows of the heatmap\n",
    "      var MapColumns = numA,\n",
    "          MapRows = numB;\n",
    "\n",
    "      //The maximum radius the hexagons can have to still fit the screen\n",
    "      var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n",
    "\n",
    "      //Calculate the center position of each hexagon\n",
    "      var points = [];\n",
    "      for (var i = 0; i < MapRows; i++) {\n",
    "          for (var j = 0; j < MapColumns; j++) {\n",
    "              var x = hexRadius * j * Math.sqrt(3)\n",
    "              //Offset each uneven row by half of a \"hex-width\" to the right\n",
    "              if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n",
    "              var y = hexRadius * i * 1.5\n",
    "              points.push([x,y])\n",
    "          }\n",
    "      }\n",
    "\n",
    "      //Set the hexagon radius\n",
    "      var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n",
    "\n",
    "      svG.append(\"g\")\n",
    "        .selectAll(\".hexagon\")\n",
    "        .data(hexbin(points))\n",
    "        .enter().append(\"path\")\n",
    "        .attr(\"class\", \"hexagon\")\n",
    "        .attr(\"d\", function (d) {\n",
    "            return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n",
    "        })\n",
    "        .attr(\"stroke\", \"white\")\n",
    "        .attr(\"stroke-width\", \"1px\")\n",
    "        .style(\"fill\", function (d,i) { return colorList[i]; })\n",
    "        .on(\"mouseover\", mover)\n",
    "        .on(\"mouseout\", mout)\n",
    "        .append(\"svg:title\")\n",
    "        .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n",
    "\n",
    "      function mover(d) {\n",
    "      d3.select(this)\n",
    "        .transition().duration(10)  \n",
    "        .style(\"fill-opacity\", 0.3)\n",
    "      }\n",
    "\n",
    "      //Mouseout function\n",
    "      function mout(d) { \n",
    "      d3.select(this)\n",
    "        .transition().duration(10)\n",
    "        .style(\"fill-opacity\", 1)\n",
    "      }\n",
    "\n",
    "    }\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({ 'chart_id': repr(chart_id) })))\n",
    "  return chart_id\n",
    "\n",
    "def Canvas(numA, numB, colorList, libraryList, indices):\n",
    "  chart_id = init_chart()\n",
    "  display(HTML(Template(dedent('''\n",
    "  <svg id=$figure_id></svg>\n",
    "  <script>\n",
    "  require([$chart_id], function(mychart) {\n",
    "    mychart($figure_id, $numA, $numB, $colorList, $libraryList, $indices)\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({\n",
    "      'chart_id': repr(chart_id),\n",
    "      'figure_id': repr('fig-' + str(uuid.uuid4())),\n",
    "      'numA': repr(numA),\n",
    "      'numB': repr(numB),\n",
    "      'colorList': repr(color_list),\n",
    "      'libraryList': repr(name),\n",
    "      'indices': repr(indices)\n",
    "  })))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "<script src=\"/static/components/requirejs/require.js\"></script>"
     },
     "metadata": {}
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "\n<script>\nrequire.config({\n  paths: {\n    'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n    'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n  },\n  shim: {\n    'd3-hexbin': ['d3']\n  }\n})\n\n// If we configure mychart via url, we can eliminate this define here\ndefine('mychart-c8de488a-1254-4458-991e-f45e76f30847', ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n  return function (figure_id, numA, numB, colorList, libraryList, indices) {\n    var margin = {top: 50, right: 20, bottom: 20, left: 50},\n      width = 850 - margin.left - margin.right,\n      height = 350 - margin.top - margin.bottom;\n\n    // append the svg object to the body of the page\n    var svG = d3.select('#' + figure_id)\n      .attr(\"width\", width + margin.left + margin.right)\n      .attr(\"height\", height + margin.top + margin.bottom)\n      .append(\"g\")\n        .attr(\"transform\",\n              \"translate(\" + margin.left + \",\" + margin.top + \")\");\n\n    //The number of columns and rows of the heatmap\n    var MapColumns = numA,\n        MapRows = numB;\n\n    //The maximum radius the hexagons can have to still fit the screen\n    var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n\n    //Calculate the center position of each hexagon\n    var points = [];\n    for (var i = 0; i < MapRows; i++) {\n        for (var j = 0; j < MapColumns; j++) {\n            var x = hexRadius * j * Math.sqrt(3)\n            //Offset each uneven row by half of a \"hex-width\" to the right\n            if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n            var y = hexRadius * i * 1.5\n            points.push([x,y])\n        }\n    }\n\n    //Set the hexagon radius\n    var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n\n    svG.append(\"g\")\n      .selectAll(\".hexagon\")\n      .data(hexbin(points))\n      .enter().append(\"path\")\n      .attr(\"class\", \"hexagon\")\n      .attr(\"d\", function (d) {\n          return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n      })\n      .attr(\"stroke\", \"white\")\n      .attr(\"stroke-width\", \"1px\")\n      .style(\"fill\", function (d,i) { return colorList[i]; })\n      .on(\"mouseover\", mover)\n      .on(\"mouseout\", mout)\n      .append(\"svg:title\")\n      .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n\n    function mover(d) {\n    d3.select(this)\n      .transition().duration(10)  \n      .style(\"fill-opacity\", 0.3)\n    }\n\n    //Mouseout function\n    function mout(d) { \n    d3.select(this)\n      .transition().duration(10)\n      .style(\"fill-opacity\", 1)\n    }\n\n  }\n})\n</script>\n"
     },
     "metadata": {}
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "\n<svg id='fig-67aead48-1659-4a03-a1c3-f921dcf2fa26'></svg>\n<script>\nrequire(['mychart-c8de488a-1254-4458-991e-f45e76f30847'], function(mychart) {\n  mychart('fig-67aead48-1659-4a03-a1c3-f921dcf2fa26', 16, 16, ['#000000', '#250025', '#000000', '#000000', '#000000', '#000000', '#220022', '#000000', '#000000', '#2f002f', '#fcba03', '#fcba03', '#170017', '#000000', '#3e003e', '#000000', '#390039', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#470047', '#000000', '#3b003b', '#000000', '#000000', '#410041', '#000000', '#490049', '#000000', '#000000', '#000000', '#000000', '#600060', '#340034', '#410041', '#2b002b', '#390039', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#330033', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#3e003e', '#360036', '#000000', '#000000', '#360036', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#790079', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#200020', '#000000', '#000000', '#000000', '#350035', '#3b003b', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#400040', '#800080', '#000000', '#000000', '#000000', '#000000', '#270027', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#400040', '#000000', '#000000', '#000000', '#000000', '#3a003a', '#000000', '#2d002d', '#000000', '#200020', '#000000', '#000000', '#000000', '#000000', '#000000', '#1e001e', '#000000', '#2f002f', '#000000', '#000000', '#000000', '#350035', '#240024', '#5b005b', '#000000', '#000000', '#000000', '#8a008a', '#000000', '#000000', '#000000', '#1b001b', '#000000', '#500050', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#400040', '#000000', '#390039', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#430043', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#340034', '#2b002b', '#8b008b', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#320032', '#000000', '#790079', '#000000', '#000000', '#470047', '#000000', '#000000', '#360036', '#000000', '#000000', '#000000', '#3e003e', '#000000', '#000000', '#110011', '#3d003d', '#290029', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#330033', '#3a003a', '#000000', '#000000', '#000000', '#000000', '#000000', '#510051', '#000000', '#000000', '#330033', '#000000', '#2d002d', '#490049', '#000000', '#470047', '#fcba03', '#fcba03', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#000000', '#fcba03', '#34ebe8', '#fcba03', '#000000', '#000000', '#000000', '#000000'], ['acetylation and deacetylation of rela in nucleus', 'actions of nitric oxide in the heart', 'activation of camp-dependent protein kinase pka', 'activation of csk by camp-dependent protein kinase inhibits signaling through the t cell receptor', 'activation of pkc through g-protein coupled receptors', 'adp-ribosylation factor', 'agrin in postsynaptic differentiation', 'akap95 role in mitosis and chromosome dynamics', 'akt signaling pathway', 'alk in cardiac myocytes', 'alternative complement pathway', 'angiotensin ii mediated activation of jnk pathway via pyk2 dependent signaling', 'antigen processing and presentation', 'apoptotic dna-fragmentation and tissue homeostasis', 'apoptotic signaling in response to dna damage', 'aspirin blocks signaling pathway involved in platelet activation', 'atm signaling pathway', 'attenuation of gpcr signaling', 'b cell survival pathway', 'basic mechanism of action of ppara pparb(d) and pparg and effects on gene expression', 'basic mechanisms of sumoylation', 'bcr signaling pathway', 'bioactive peptide induced signaling pathway', 'bone remodeling', 'brca1 dependent ub ligase activity', 'btg family proteins and cell cycle regulation', 'ca-calmodulin-dependent protein kinase activation', 'cadmium induces dna synthesis and proliferation in macrophages', 'calcium signaling by hbx of hepatitis b virus', 'cardiac protection against ros', 'carm1 and regulation of the estrogen receptor', 'caspase cascade in apoptosis', 'cbl mediated ligand-induced downregulation of egf receptors pathway', 'ccr3 signaling in eosinophils', 'cd40l signaling pathway', 'cdc25 and chk1 regulatory pathway in response to dna damage', 'cdk regulation of dna replication', 'cell cycle: g1/s check point', 'cell cycle: g2/m checkpoint', 'cell to cell adhesion signaling', 'ceramide signaling pathway', 'chaperones modulate interferon signaling pathway', 'chrebp regulation by carbohydrates and camp', 'chromatin remodeling by hswi/snf atp-dependent complexes', 'classical complement pathway', 'control of gene expression by vitamin d receptor', 'control of skeletal myogenesis by hdac and calcium/calmodulin-dependent kinase (camk)', 'corticosteroids and cardioprotection', 'ctcf: first multivalent nuclear factor', 'cxcr4 signaling pathway', 'cyclin e destruction pathway', 'cycling of ran in nucleocytoplasmic transport', 'cyclins and cell cycle regulation', 'cystic fibrosis transmembrane conductance regulator (cftr) and beta 2 adrenergic receptor (b2ar) pathway', 'd4gdi signaling pathway', 'deregulation of cdk5 in alzheimers disease', 'double stranded rna induced gene expression', 'downregulated of mta-3 in er-negative breast tumors', 'e2f1 destruction pathway', 'effects of calcineurin in keratinocyte differentiation', 'egf signaling pathway', 'eicosanoid metabolism', 'endocytotic role of ndk phosphins and dynamin', 'epo signaling pathway', 'er associated degradation (erad) pathway', 'erk and pi-3 kinase are necessary for collagen binding in corneal epithelia', 'erk1/erk2 mapk signaling pathway', 'erythropoietin mediated neuroprotection through nf-kb', 'estrogen responsive protein efp controls cell cycle and breast tumors growth', 'eukaryotic protein translation', 'extrinsic prothrombin activation pathway', 'fas signaling pathway (cd95)', 'fc epsilon receptor i signaling in mast cells', 'fibrinolysis pathway', 'fmlp induced chemokine gene expression in hmc-1 cells', 'fosb gene expression and drug abuse', 'g-protein signaling through tubby proteins', 'g-secretase mediated erbb4 signaling pathway', 'gamma-aminobutyric acid receptor life cycle pathway', 'gata3 participate in activating the th2 cytokine genes expression', 'granzyme a mediated apoptosis pathway', 'growth hormone signaling pathway', 'hemoglobins chaperone', 'hiv-1 nef: negative effector of fas and tnf', 'how does salmonella hijack a cell', 'how progesterone initiates the oocyte maturation', 'human cytomegalovirus and map kinase pathways', 'hypoxia and p53 in the cardiovascular system', 'hypoxia-inducible factor in the cardivascular system', 'ifn alpha signaling pathway', 'ifn gamma signaling pathway', 'igf-1 signaling pathway', 'il 2 signaling pathway', 'il 3 signaling pathway', 'il 4 signaling pathway', 'il 6 signaling pathway', 'il-10 anti-inflammatory signaling pathway', 'il-2 receptor beta chain in t cell activation', 'il-7 signal transduction', 'il12 and stat4 dependent signaling pathway in th1 development', 'il22 soluble receptor signaling pathway', 'inactivation of gsk3 by akt causes accumulation of b-catenin in alveolar macrophages', 'induction of apoptosis through dr3 and dr4/5 death receptors', 'influence of ras and rho proteins on g1 to s transition', 'inhibition of cellular proliferation by gleevec', 'inhibition of matrix metalloproteinases', 'insulin signaling pathway', 'integrin signaling pathway', 'internal ribosome entry pathway', 'intrinsic prothrombin activation pathway', 'ion channels and their functional role in vascular endothelium', 'keratinocyte differentiation', 'lck and fyn tyrosine kinases in initiation of tcr activation', 'lectin induced complement pathway', 'links between pyk2 and map kinases', 'lissencephaly gene (lis1) in neuronal migration and development', 'map kinase inactivation of smrt corepressor', 'mapkinase signaling pathway', 'mcalpain and friends in cell motility', 'mechanism of acetaminophen activity and toxicity', 'mechanism of gene regulation by peroxisome proliferators via ppara', 'mechanism of protein import into the nucleus', 'mechanisms of transcriptional repression by dna methylation', 'melanocyte development and pigmentation pathway', 'metabolism of anandamide an endogenous cannabinoid', 'mets affect on macrophage differentiation', 'mtor signaling pathway', 'multi-drug resistance factors', 'multi-step regulation of transcription by pitx2', 'multiple antiapoptotic pathways from igf-1r signaling lead to bad phosphorylation', 'nerve growth factor pathway (ngf)', 'neuroregulin receptor degredation protein-1 controls erbb3 receptor recycling', 'nf-kb signaling pathway', 'nfat and hypertrophy of the heart ', 'nfkb activation by nontypeable hemophilus influenzae', 'nitric oxide signaling pathway', 'no2-dependent il-12 pathway in nk cells', 'nuclear receptors coordinate the activities of chromatin remodeling complexes and coactivators to facilitate initiation of transcription in carcinoma cells', 'overview of telomerase protein component gene htert transcriptional regulation', 'overview of telomerase rna component gene hterc transcriptional regulation', 'oxidative stress induced gene expression via nrf2', 'p38 mapk signaling pathway', 'p53 signaling pathway', 'pdgf signaling pathway', 'pelp1 modulation of estrogen receptor activity', 'pertussis toxin-insensitive ccr5 signaling in macrophage', 'phosphoinositides and their downstream targets', 'phospholipase c signaling pathway', 'phospholipase c-epsilon pathway', 'phospholipids as signalling intermediaries', 'phosphorylation of mek1 by cdk5/p35 down regulates the map kinase pathway', 'pkc-catalyzed phosphorylation of inhibitory phosphoprotein of myosin phosphatase', 'platelet amyloid precursor protein pathway', 'polyadenylation of mrna', 'prion pathway', 'proteasome complex', 'protein kinase a at the centrosome', 'proteolysis and signaling pathway of notch', 'pten dependent cell cycle arrest and apoptosis', 'rac1 cell motility signaling pathway', 'ras signaling pathway', 'ras-independent pathway in nk cell-mediated cytotoxicity', 'rb tumor suppressor/checkpoint signaling in response to dna damage', 'regulation of bad phosphorylation', 'regulation of cell cycle progression by plk3', 'regulation of ck1/cdk5 by type 1 glutamate receptors', 'regulation of eif-4e and p70s6 kinase', 'regulation of eif2', 'regulation of map kinase pathways through dual specificity phosphatases', 'regulation of p27 phosphorylation during cell cycle progression', 'regulation of pgc-1a', 'regulation of spermatogenesis by crem', 'regulation of splicing through sam68', 'regulation of transcriptional activity by pml', 'regulators of bone mineralization', 'repression of pain sensation by the transcriptional regulator dream', 'reversal of insulin resistance by leptin', 'rho cell motility signaling pathway', 'rho-selective guanine exchange factor akap13 mediates stress fiber formation', 'rna polymerase iii transcription', 'role of brca1 brca2 and atr in cancer susceptibility', 'role of egf receptor transactivation by gpcrs in cardiac hypertrophy', 'role of erbb2 in signal transduction and oncology', 'role of erk5 in neuronal survival pathway', 'role of mal in rho-mediated activation of srf', 'role of mef2d in t-cell apoptosis', 'role of mitochondria in apoptotic signaling', 'role of nicotinic acetylcholine receptors in the regulation of apoptosis', 'role of parkin in ubiquitin-proteasomal pathway', 'role of pi3k subunit p85 in regulation of actin organization and cell migration', 'role of ppar-gamma coactivators in obesity and thermogenesis', 'role of ran in mitotic spindle regulation', 'role of ﬂ-arrestins in the activation and targeting of map kinases', 'roles of ﬂ arrestin dependent recruitment of src kinases in gpcr signaling', 'segmentation clock', 'signal dependent regulation of myogenesis by corepressor mitr', 'signal transduction through il1r', 'signaling pathway from g-protein families', 'skeletal muscle hypertrophy is regulated via akt-mtor pathway', 'sodd/tnfr1 signaling pathway', 'sonic hedgehog receptor ptc1 regulates cell cycle', 'spliceosomal assembly', 'sprouty regulation of tyrosine kinase signals', 'srebp control of lipid synthesis', 'stat3 signaling pathway', 'stathmin and breast cancer resistance to antimicrotubule agents', 'stress induction of hsp regulation', 'sumoylation as a mechanism to modulate ctbp-dependent gene responses', 'sumoylation by ranbp2 regulates transcriptional repression', 't cell receptor signaling pathway', 'telomeres telomerase cellular aging and immortality', 'tgf beta signaling pathway', 'the 41bb-dependent immune response', 'the co-stimulatory signal during t-cell activation', 'the igf-1 receptor and longevity', 'the information processing pathway at the ifn beta enhancer', 'the prc2 complex sets long-term gene silencing through modification of histone tails', 'thrombin signaling and protease-activated receptors', 'tnf/stress related signaling', 'tnfr1 signaling pathway', 'tnfr2 signaling pathway', 'toll-like receptor pathway', 'tpo signaling pathway', 'transcription factor creb and its extracellular signals', 'transcription regulation by methyltransferase of carm1', 'trefoil factors initiate mucosal healing', 'trka receptor signaling pathway', 'tsp-1 induced apoptosis in microvascular endothelial cell', 'tumor suppressor arf inhibits ribosomal biogenesis', 'ucalpain and friends in cell spread', 'vegf hypoxia and angiogenesis', 'visceral fat deposits and the metabolic syndrome', 'visual signal transduction', 'west nile virus', 'wnt lrp6 signalling', 'wnt signaling pathway', 'y branching of actin filaments', 'yaci and bcma stimulation of b cell immune responses', 'ﬂ-arrestins in gpcr desensitization'], [0.0, 0.01449, 0.0, 0.0, 0.0, 0.0, 0.01351, 0.0, 0.0, 0.01852, 0.0, 0.01639, 0.00885, 0.0, 0.02439, 0.0, 0.02222, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02778, 0.0, 0.02326, 0.0, 0.0, 0.02564, 0.0, 0.02857, 0.0, 0.0, 0.0, 0.0, 0.03774, 0.02041, 0.02564, 0.01667, 0.02222, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02439, 0.02128, 0.0, 0.0, 0.02128, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.04762, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0125, 0.0, 0.0, 0.0, 0.02083, 0.02326, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.025, 0.05, 0.0, 0.0, 0.0, 0.0, 0.01515, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.025, 0.0, 0.0, 0.0, 0.0, 0.02273, 0.0, 0.01754, 0.0, 0.01266, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0119, 0.0, 0.01852, 0.0, 0.0, 0.0, 0.02083, 0.01408, 0.03571, 0.0, 0.0, 0.0, 0.05405, 0.0, 0.0, 0.0, 0.01042, 0.0, 0.03125, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.025, 0.0, 0.02222, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02632, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02041, 0.01667, 0.05455, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01961, 0.0, 0.04762, 0.0, 0.0, 0.02778, 0.0, 0.0, 0.02128, 0.0, 0.0, 0.0, 0.02439, 0.0, 0.0, 0.00667, 0.02381, 0.01613, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.02, 0.02273, 0.0, 0.0, 0.0, 0.0, 0.0, 0.03175, 0.0, 0.0, 0.02, 0.0, 0.01754, 0.02857, 0.0, 0.02778, 0.0, 0.01111, 0.0, 0.0, 0.0])\n})\n</script>\n"
     },
     "metadata": {}
    }
   ],
   "source": [
    "Canvas(x_dimension, y_dimension, color_list, name, indices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('base': conda)",
   "language": "python",
   "name": "python37664bitbasecondad81f4d096e8b4b9885e3d702e0bb3be1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}