{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Canvas Creation from Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import json\n",
    "import requests\n",
    "import math\n",
    "import matplotlib\n",
    "import uuid\n",
    "import urllib\n",
    "from textwrap import dedent\n",
    "from IPython.core.display import display, HTML\n",
    "from string import Template\n",
    "from random import seed, randint\n",
    "from operator import itemgetter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below this is arbitrary definitions for these variables, their actual values will be selected in the Appyter (so DON'T ADD BELOW CELL TO APPYTER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_libraries = ['WikiPathways_2019_Human']\n",
    "color = 'Purple'\n",
    "scaling_factor = ''\n",
    "num_hex_colored = '20'\n",
    "\n",
    "# list of genes that was input\n",
    "\n",
    "open_gene_list_file = open('geneList.txt','r')\n",
    "lines = open_gene_list_file.readlines()\n",
    "genes = [x.strip().upper() for x in lines]\n",
    "open_gene_list_file.close()\n",
    "\n",
    "#genes = ['TP53', 'TNF', 'EGFR', 'GKN1', 'HADHA', 'APOE', 'ESR1', 'VEGFA', 'TGFB1', 'PREPL', 'TIA1', 'TPO', 'TTN', 'SATB2', 'CHPF', 'MALL', 'MIPIP', 'NUPL1', 'IL6', 'PDIA3', 'CTNNB1', 'SLC39A1', 'DTNA','SLC1A1', 'GALNT2', 'HIST2H2AC', 'CD63']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def library_processing(library_index):\n",
    "    # processes library data\n",
    "\n",
    "    raw_library_data = []\n",
    "    library_data = []\n",
    "    library_name = all_libraries[library_index]\n",
    "\n",
    "    # pull library from online\n",
    "\n",
    "    with urllib.request.urlopen('https://raw.githubusercontent.com/skylar73/Enrichr-Library-Annealing/master/Annealed-Libraries/' + all_libraries[library_index] + '.txt') as f:\n",
    "        for line in f.readlines():\n",
    "            raw_library_data.append(line.decode('utf-8').split(\"\\t\\t\"))\n",
    "\n",
    "    # OR\n",
    "    # locally upload library\n",
    "\n",
    "    # with open('filepath', 'r') as f:\n",
    "    #     for line in f.readlines():\n",
    "    #        raw_library_data.append(line.split(\"\\t\\t\"))\n",
    "\n",
    "\n",
    "    name = []\n",
    "    gene_list = []\n",
    "\n",
    "    for i in range(len(raw_library_data)):\n",
    "        name += [raw_library_data[i][0]]\n",
    "        raw_genes = raw_library_data[i][1].split('\\t')\n",
    "        gene_list += [raw_genes[:-1]]\n",
    "\n",
    "    library_data = [list(a) for a in zip(name, gene_list)]\n",
    "\n",
    "    # raw_library_data: a 2D list where the first element is the name and the second element is a list of genes associated with that name\n",
    "\n",
    "    jaccard_indices = []\n",
    "    indices = []\n",
    "\n",
    "    for gene_set in library_data:\n",
    "        intersection = [value for value in gene_set[1] if value in genes]\n",
    "        index = len(intersection)/(len(gene_set[1]) + len(genes))\n",
    "        jaccard_indices += [[gene_set[0], index]]\n",
    "        indices += [round(index, 5)]\n",
    "\n",
    "    # determine the dimensions of the canvas\n",
    "    x_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "    y_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "\n",
    "    # zip name, gene_list, indices, and blank list for neighbor score then add dummy entries to the zipped list\n",
    "    num_hex = x_dimension*y_dimension\n",
    "    anneal_list = list(zip(name, gene_list, indices))\n",
    "\n",
    "    return anneal_list, x_dimension, y_dimension"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "def unzip_list(anneal_list):\n",
    "    unzipped_list = zip(*anneal_list)\n",
    "    return list(unzipped_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_color(anneal_list, cut_off_value, scaling_factor, x_dimension, y_dimension):\n",
    "\n",
    "    # Deal with cut_off_value (only color the most significant 10/20 hexagons)\n",
    "    if cut_off_value == 2.0:\n",
    "        sort_list = sorted(anneal_list, key=itemgetter(2), reverse=True)\n",
    "        cut_off_value = sort_list[int(num_hex_colored)-1][2]\n",
    "\n",
    "    r_value = 0\n",
    "    g_value = 0\n",
    "    b_value = 0\n",
    "\n",
    "    if color == 'Red':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.8\n",
    "        b_value = 0.8\n",
    "    if color == 'Yellow':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.3\n",
    "        b_value = 1.0\n",
    "    if color == 'Purple':\n",
    "        r_value = 0.5\n",
    "        g_value = 1.0\n",
    "        b_value = 0.0\n",
    "    if color == 'Pink':\n",
    "        r_value = 0.0\n",
    "        g_value = 1.0\n",
    "        b_value = 0.2\n",
    "    if color == 'Orange':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.45\n",
    "        b_value = 1.0\n",
    "    if color == 'Green':\n",
    "        r_value = 1.0\n",
    "        g_value = 0.0\n",
    "        b_value = 1.0\n",
    "    if color == 'Blue':\n",
    "        r_value = 1.0\n",
    "        g_value = 0.9\n",
    "        b_value = 0.0\n",
    "\n",
    "    color_list = []\n",
    "\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "\n",
    "    if scaling_factor == '':\n",
    "        max_index = max(unzipped_anneal_list[2])\n",
    "    else:\n",
    "        max_index = float(scaling_factor)\n",
    "\n",
    "    scaled_list = [i/max_index for i in unzipped_anneal_list[2]]\n",
    "\n",
    "    for i in range(x_dimension*y_dimension):\n",
    "        if i < len(unzipped_anneal_list[2]) and float(unzipped_anneal_list[2][i]) >= cut_off_value:\n",
    "            color_list += [matplotlib.colors.to_hex((1-scaled_list[i]*r_value, \n",
    "            1-scaled_list[i]*g_value, 1-scaled_list[i]*b_value))]\n",
    "        elif i < len(unzipped_anneal_list[2]):\n",
    "            color_list += [matplotlib.colors.to_hex((1-scaled_list[i], \n",
    "            1-scaled_list[i], 1-scaled_list[i]))]\n",
    "        else:\n",
    "            color_list += [\"#FFFFFF\"]\n",
    "    return color_list, max_index, cut_off_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "def init_chart():\n",
    "  chart_id = 'mychart-' + str(uuid.uuid4())\n",
    "  display(HTML('<script src=\"/static/components/requirejs/require.js\">'))\n",
    "  display(HTML(Template(dedent('''\n",
    "  <script>\n",
    "  require.config({\n",
    "    paths: {\n",
    "      'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n",
    "      'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n",
    "    },\n",
    "    shim: {\n",
    "      'd3-hexbin': ['d3']\n",
    "    }\n",
    "  })\n",
    "\n",
    "  // If we configure mychart via url, we can eliminate this define here\n",
    "  define($chart_id, ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n",
    "    return function (figure_id, numA, numB, colorList, libraryList, indices) {\n",
    "      var margin = {top: 50, right: 20, bottom: 20, left: 50},\n",
    "        width = 850 - margin.left - margin.right,\n",
    "        height = 350 - margin.top - margin.bottom;\n",
    "\n",
    "      // append the svg object to the body of the page\n",
    "      var svG = d3.select('#' + figure_id)\n",
    "                  .attr(\"width\", width + margin.left + margin.right)\n",
    "                  .attr(\"height\", height + margin.top + margin.bottom)\n",
    "                  .append(\"g\")\n",
    "                  .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
    "      \n",
    "      //The number of columns and rows of the heatmap\n",
    "      var MapColumns = numA,\n",
    "          MapRows = numB;\n",
    "\n",
    "      //The maximum radius the hexagons can have to still fit the screen\n",
    "      var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n",
    "\n",
    "      //Calculate the center position of each hexagon\n",
    "      var points = [];\n",
    "      for (var i = 0; i < MapRows; i++) {\n",
    "          for (var j = 0; j < MapColumns; j++) {\n",
    "              var x = hexRadius * j * Math.sqrt(3)\n",
    "              //Offset each uneven row by half of a \"hex-width\" to the right\n",
    "              if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n",
    "              var y = hexRadius * i * 1.5\n",
    "              points.push([x,y])\n",
    "          }\n",
    "      }\n",
    "\n",
    "      //Set the hexagon radius\n",
    "      var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n",
    "\n",
    "      svG.append(\"g\")\n",
    "        .selectAll(\".hexagon\")\n",
    "        .data(hexbin(points))\n",
    "        .enter().append(\"path\")\n",
    "        .attr(\"class\", \"hexagon\")\n",
    "        .attr(\"d\", function (d) {\n",
    "            return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n",
    "        })\n",
    "        .attr(\"stroke\", \"black\")\n",
    "        .attr(\"stroke-width\", \"1px\")\n",
    "        .style(\"fill\", function (d,i) { return colorList[i]; })\n",
    "        .on(\"mouseover\", mover)\n",
    "        .on(\"mouseout\", mout)\n",
    "        .append(\"svg:title\")\n",
    "        .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n",
    "\n",
    "      // Mouseover function\n",
    "      function mover(d) {\n",
    "      d3.select(this)\n",
    "        .transition().duration(10)  \n",
    "        .style(\"fill-opacity\", 0.3)\n",
    "      };\n",
    "\n",
    "      // Mouseout function\n",
    "      function mout(d) { \n",
    "      d3.select(this)\n",
    "        .transition().duration(10)\n",
    "        .style(\"fill-opacity\", 1)\n",
    "      };\n",
    "\n",
    "  }\n",
    "\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({ 'chart_id': repr(chart_id) })))\n",
    "  return chart_id\n",
    "\n",
    "def Canvas(numA, numB, colorList, libraryList, indices):\n",
    "  chart_id = init_chart()\n",
    "  display(HTML(Template(dedent('''\n",
    "  <svg id=$figure_id></svg>\n",
    "  <script>\n",
    "  require([$chart_id], function(mychart) {\n",
    "    mychart($figure_id, $numA, $numB, $colorList, $libraryList, $indices)\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({\n",
    "      'chart_id': repr(chart_id),\n",
    "      'figure_id': repr('fig-' + str(uuid.uuid4())),\n",
    "      'numA': repr(numA),\n",
    "      'numB': repr(numB),\n",
    "      'colorList': repr(colorList),\n",
    "      'libraryList': repr(libraryList),\n",
    "      'indices': repr(indices)\n",
    "  })))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "<script src=\"/static/components/requirejs/require.js\">"
     },
     "metadata": {}
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "\n<script>\nrequire.config({\n  paths: {\n    'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n    'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n  },\n  shim: {\n    'd3-hexbin': ['d3']\n  }\n})\n\n// If we configure mychart via url, we can eliminate this define here\ndefine('mychart-ee9741b8-3f95-4976-8244-8483c6bc7925', ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n  return function (figure_id, numA, numB, colorList, libraryList, indices) {\n    var margin = {top: 50, right: 20, bottom: 20, left: 50},\n      width = 850 - margin.left - margin.right,\n      height = 350 - margin.top - margin.bottom;\n\n    // append the svg object to the body of the page\n    var svG = d3.select('#' + figure_id)\n                .attr(\"width\", width + margin.left + margin.right)\n                .attr(\"height\", height + margin.top + margin.bottom)\n                .append(\"g\")\n                .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n\n    //The number of columns and rows of the heatmap\n    var MapColumns = numA,\n        MapRows = numB;\n\n    //The maximum radius the hexagons can have to still fit the screen\n    var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n\n    //Calculate the center position of each hexagon\n    var points = [];\n    for (var i = 0; i < MapRows; i++) {\n        for (var j = 0; j < MapColumns; j++) {\n            var x = hexRadius * j * Math.sqrt(3)\n            //Offset each uneven row by half of a \"hex-width\" to the right\n            if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n            var y = hexRadius * i * 1.5\n            points.push([x,y])\n        }\n    }\n\n    //Set the hexagon radius\n    var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n\n    svG.append(\"g\")\n      .selectAll(\".hexagon\")\n      .data(hexbin(points))\n      .enter().append(\"path\")\n      .attr(\"class\", \"hexagon\")\n      .attr(\"d\", function (d) {\n          return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n      })\n      .attr(\"stroke\", \"black\")\n      .attr(\"stroke-width\", \"1px\")\n      .style(\"fill\", function (d,i) { return colorList[i]; })\n      .on(\"mouseover\", mover)\n      .on(\"mouseout\", mout)\n      .append(\"svg:title\")\n      .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n\n    // Mouseover function\n    function mover(d) {\n    d3.select(this)\n      .transition().duration(10)  \n      .style(\"fill-opacity\", 0.3)\n    };\n\n    // Mouseout function\n    function mout(d) { \n    d3.select(this)\n      .transition().duration(10)\n      .style(\"fill-opacity\", 1)\n    };\n\n}\n\n})\n</script>\n"
     },
     "metadata": {}
    },
    {
     "output_type": "display_data",
     "data": {
      "text/plain": "<IPython.core.display.HTML object>",
      "text/html": "\n<svg id='fig-68c7835b-390d-42f6-b7f8-10b27b87fd14'></svg>\n<script>\nrequire(['mychart-ee9741b8-3f95-4976-8244-8483c6bc7925'], function(mychart) {\n  mychart('fig-68c7835b-390d-42f6-b7f8-10b27b87fd14', 22, 22, ['#c7c7c7', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#b264ff', '#ffffff', '#c7c7c7', '#959595', '#ffffff', '#ffffff', '#ffffff', '#c1c1c1', '#c1c1c1', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c4c4c4', '#b266ff', '#ffffff', '#c3c3c3', '#c3c3c3', '#ffffff', '#ffffff', '#8f8f8f', '#c9c9c9', '#c6c6c6', '#ffffff', '#cbcbcb', '#cccccc', '#ffffff', '#ffffff', '#ae5cff', '#d2d2d2', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c4c4c4', '#ffffff', '#838383', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c2c2c2', '#ffffff', '#ffffff', '#c3c3c3', '#c8c8c8', '#797979', '#ffffff', '#ffffff', '#ffffff', '#878787', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#989898', '#ffffff', '#ffffff', '#b46aff', '#929292', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#c6c6c6', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c5c5c5', '#ffffff', '#c5c5c5', '#9e9e9e', '#a5a5a5', '#ac59ff', '#a5a5a5', '#ac5aff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c7c7c7', '#ffffff', '#ffffff', '#ffffff', '#cccccc', '#c8c8c8', '#ffffff', '#a4a4a4', '#d1d1d1', '#c9c9c9', '#ffffff', '#b66eff', '#9a9a9a', '#cdcdcd', '#a4a4a4', '#a750ff', '#af5eff', '#c1c1c1', '#ffffff', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c2c2c2', '#c4c4c4', '#c5c5c5', '#ffffff', '#797979', '#ffffff', '#ffffff', '#9a9a9a', '#c4c4c4', '#c5c5c5', '#ffffff', '#ffffff', '#c9c9c9', '#a5a5a5', '#cccccc', '#ffffff', '#ffffff', '#ffffff', '#c9c9c9', '#ffffff', '#868686', '#ffffff', '#8e8e8e', '#ffffff', '#868686', '#b56bff', '#ffffff', '#ffffff', '#c6c6c6', '#ffffff', '#ffffff', '#ffffff', '#c5c5c5', '#c1c1c1', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c6c6c6', '#c6c6c6', '#c9c9c9', '#c6c6c6', '#c7c7c7', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#9d9d9d', '#cccccc', '#cdcdcd', '#828282', '#c8c8c8', '#c8c8c8', '#ffffff', '#cacaca', '#cacaca', '#c6c6c6', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#838383', '#ffffff', '#ffffff', '#7a7a7a', '#a64dff', '#d2d2d2', '#ffffff', '#ffffff', '#909090', '#cccccc', '#b163ff', '#c4c4c4', '#c6c6c6', '#c5c5c5', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#c2c2c2', '#8a8a8a', '#c5c5c5', '#c1c1c1', '#a346ff', '#909090', '#ffffff', '#c4c4c4', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#cccccc', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#ffffff', '#c6c6c6', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#c2c2c2', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#b264ff', '#c6c6c6', '#ffffff', '#c5c5c5', '#858585', '#cacaca', '#6f6f6f', '#b468ff', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c2c2c2', '#ffffff', '#949494', '#888888', '#ffffff', '#ffffff', '#cacaca', '#c5c5c5', '#8e8e8e', '#a54aff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#949494', '#c4c4c4', '#ffffff', '#c9c9c9', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#929292', '#c2c2c2', '#ffffff', '#ffffff', '#c4c4c4', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#7f7f7f', '#d0d0d0', '#cccccc', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#cbcbcb', '#767676', '#ffffff', '#ffffff', '#cccccc', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c2c2c2', '#ffffff', '#8c8c8c', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c4c4c4', '#ffffff', '#a0a0a0', '#c5c5c5', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#aa55ff', '#ffffff', '#c7c7c7', '#c7c7c7', '#ffffff', '#a1a1a1', '#ffffff', '#ffffff', '#c6c6c6', '#ffffff', '#ffffff', '#ffffff', '#c1c1c1', '#c5c5c5', '#ffffff', '#ffffff', '#8000ff', '#aa54ff', '#838383', '#858585', '#ffffff', '#ffffff', '#ffffff', '#898989', '#ffffff', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#a4a4a4', '#ffffff', '#c3c3c3', '#c1c1c1', '#ffffff', '#727272', '#c5c5c5', '#ffffff', '#ffffff', '#cccccc', '#ffffff', '#8e8e8e', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#9c9c9c', '#9a36ff', '#ffffff', '#cccccc', '#ffffff', '#ffffff', '#c9c9c9', '#898989', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#ffffff', '#cccccc', '#d1d1d1', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#858585', '#c5c5c5', '#cbcbcb', '#ffffff', '#ffffff', '#c6c6c6', '#ffffff', '#ffffff', '#c6c6c6', '#ffffff', '#c4c4c4', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c1c1c1', '#c2c2c2', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#c8c8c8'], ['Structural Pathway of Interleukin 1 (IL-1) WP2637', 'Signal transduction through IL1R WP4496', 'RANKL/RANK (Receptor activator of NFKB (ligand)) Signaling Pathway WP2018', 'TP53 Network WP1742', 'Metastatic brain tumor WP2249', 'Composition of Lipid Particles WP3601', 'Statin Pathway WP430', 'Selenium Micronutrient Network WP15', 'IL1 and megakaryocytes in obesity WP2865', 'Cardiac Progenitor Differentiation WP2406', 'Primary Focal Segmental Glomerulosclerosis FSGS WP2572', 'Mammary gland development pathway - Embryonic development (Stage 1 of 4) WP2813', 'Regulation of Apoptosis by Parathyroid Hormone-related Protein WP3872', 'Mechanoregulation and pathology of YAP/TAZ via Hippo and non-Hippo mechanisms WP4534', 'Steroid Biosynthesis WP496', 'The alternative pathway of fetal androgen synthesis WP4524', 'Oxidative Stress WP408', 'H19 action Rb-E2F1 signaling and CDK-Beta-catenin activity WP3969', 'Mammary gland development pathway - Involution (Stage 4 of 4) WP2815', 'Irinotecan Pathway WP229', 'Glycosylation and related congenital defects. WP4521', 'Cannabinoid receptor signaling WP3869', 'MAPK Signaling Pathway WP382', 'TNF related weak inducer of apoptosis (TWEAK) Signaling Pathway WP2036', 'Estrogen signaling pathway WP712', 'miRNA regulation of p53 pathway in prostate cancer WP3982', 'ApoE and miR-146 in inflammation and atherosclerosis WP3926', 'Metabolic pathway of LDL, HDL and TG, including diseases WP4522', 'Vitamin B12 Metabolism WP1533', 'Folate Metabolism WP176', 'Selenium Metabolism and Selenoproteins WP28', 'Differentiation Pathway WP2848', 'Viral Acute Myocarditis WP4298', 'Apoptosis Modulation and Signaling WP1772', 'Apoptosis Modulation by HSP70 WP384', 'Alzheimers Disease WP2059', 'Genotoxicity pathway WP4286', 'Integrated Breast Cancer Pathway WP1984', 'Integrated Cancer Pathway WP1971', 'Prader-Willi and Angelman Syndrome WP3998', 'let-7 inhibition of ES cell reprogramming WP3299', 'Phase I biotransformations, non P450 WP136', 'Small Ligand GPCRs WP247', 'Signal Transduction of S1P Receptor WP26', 'Photodynamic therapy-induced unfolded protein response WP3613', 'NLR Proteins WP288', 'Circadian rhythm related genes WP3594', 'Exercise-induced Circadian Regulation WP410', 'Synthesis and Degradation of Ketone Bodies WP311', '', 'Cells and Molecules involved in local acute inflammatory response  WP4493', 'Photodynamic therapy-induced NF-kB survival signaling WP3617', 'Resistin as a regulator of inflammation WP4481', 'ncRNAs involved in STAT3 signaling in hepatocellular carcinoma WP4337', 'Cell Differentiation - Index WP2029', 'Cell Differentiation - Index expanded WP2023', 'Nanomaterial induced apoptosis WP2507', 'Fas Ligand (FasL) pathway and Stress induction of Heat Shock Proteins (HSP) regulation WP314', 'Imatinib and Chronic Myeloid Leukemia WP3640', 'ATR Signaling WP3875', 'DNA Mismatch Repair WP531', 'Cell Cycle WP179', 'G1 to S cell cycle control WP45', 'Kennedy pathway from Sphingolipids WP3933', 'GPR40 Pathway WP3958', 'BMP2-WNT4-FOXO1 Pathway in Human Primary Endometrial Stromal Cell Differentiation WP3876', 'Complement Activation WP545', 'Complement and Coagulation Cascades WP558', 'Sudden Infant Death Syndrome (SIDS) Susceptibility Pathways WP706', 'GPCRs, Class B Secretin-like WP334', '', 'Catalytic cycle of mammalian Flavin-containing MonoOxygenases (FMOs) WP688', 'Vitamin D in inflammatory diseases WP4482', 'Role Altered Glycolysation of MUC1 in Tumour Microenvironment WP4480', 'Supression of HMGB1 mediated inflammation by THBD WP4479', 'IL17 signaling pathway WP2112', 'TFs Regulate miRNAs related to cardiac hypertrophy WP1559', 'SRF and miRs in Smooth Muscle Differentiation and Proliferation WP1991', 'DNA Damage Response WP707', 'miRNA Regulation of DNA Damage Response WP1530', 'DNA IR-Double Strand Breaks (DSBs) and cellular response via ATM WP3959', 'DNA IR-damage and cellular response via ATR WP4016', 'Homologous recombination WP186', 'DNA Replication WP466', 'Retinoblastoma Gene in Cancer WP2446', 'Glycerophospholipid Biosynthetic Pathway WP2533', 'PTF1A related regulatory pathway WP4147', 'Role of Osx and miRNAs in tooth development WP3971', 'ncRNAs involved in Wnt signaling in hepatocellular carcinoma WP4336', 'Blood Clotting Cascade WP272', 'MECP2 and Associated Rett Syndrome WP3584', 'Electron Transport Chain (OXPHOS system in mitochondria) WP111', 'Oxidative phosphorylation WP623', 'Nicotine Metabolism WP1600', 'Arylamine metabolism WP694', 'MAPK  and NFkB Signalling Pathways Inhibited by Yersinia YopJ WP3849', 'EBV LMP1 signaling WP262', 'Osteopontin Signaling WP1434', 'Extracellular vesicles in the crosstalk of cardiac cells WP4300', 'miRNA Biogenesis WP2338', 'DDX1 as a regulatory component of the Drosha microprocessor WP2942', 'ATM Signaling Pathway WP2516', 'ATM Signaling Network in Development and Disease  WP3878', 'Genes targeted by miRNAs in adipocytes WP1992', 'Ethanol metabolism resulting in production of ROS by CYP2E1 WP4269', 'Regulation of sister chromatid separation at the metaphase-anaphase transition WP4240', 'The effect of progerin on the involved genes in Hutchinson-Gilford Progeria Syndrome WP4320', 'Tumor suppressor activity of SMARCB1 WP4204', 'Gene regulatory network modelling somitogenesis  WP2854', 'Wnt Signaling in Kidney Disease WP4150', 'Wnt Signaling WP428', 'Myometrial Relaxation and Contraction Pathways WP289', 'PI3K-Akt Signaling Pathway WP4172', 'Nonalcoholic fatty liver disease WP4396', 'Mitochondrial complex I assembly model OXPHOS system WP4324', 'Tamoxifen metabolism WP691', 'Estrogen metabolism WP697', 'Codeine and Morphine Metabolism WP1604', 'TLR4 Signaling and Tolerance WP3851', 'Apoptosis-related network due to altered Notch3 in ovarian cancer WP2864', '', 'exRNA mechanism of action and biogenesis WP2805', 'Non-homologous end joining WP438', 'GPCRs, Other WP117', 'Hematopoietic Stem Cell Differentiation WP2849', 'Development of pulmonary dendritic cells and macrophage subsets WP3892', 'Mesodermal Commitment Pathway WP2857', 'Endoderm Differentiation WP2853', 'Histone Modifications WP2369', 'Interactome of polycomb repressive complex 2 (PRC2)  WP2916', 'ESC Pluripotency Pathways WP3931', 'LncRNA involvement in canonical Wnt signaling and colorectal cancer WP4258', 'Wnt Signaling Pathway and Pluripotency WP399', 'Calcium Regulation in the Cardiac Cell WP536', 'Focal Adhesion-PI3K-Akt-mTOR-signaling pathway WP3932', 'AMP-activated Protein Kinase (AMPK) Signaling WP1403', 'Leptin and adiponectin WP3934', 'Fatty Acid Omega Oxidation WP206', 'Benzo(a)pyrene metabolism WP696', 'Zinc homeostasis WP3529', 'Copper homeostasis WP3286', 'Senescence and Autophagy in Cancer WP615', 'Nanoparticle triggered autophagic cell death WP2509', 'ACE Inhibitor Pathway WP554', 'Peptide GPCRs WP24', 'GPCRs, Class A Rhodopsin-like WP455', '', 'Tgif disruption of Shh signaling WP3674', 'EDA Signalling in Hair Follicle Development WP3930', 'Hedgehog Signaling Pathway WP47', 'Gastric Cancer Network 1 WP2361', 'Fatty Acid Beta Oxidation WP143', 'Fatty Acid Biosynthesis WP357', 'Epithelial to mesenchymal transition in colorectal cancer WP4239', 'NO/cGMP/PKG mediated Neuroprotection WP4008', 'Phosphodiesterases in neuronal function WP4222', 'Human Complement System WP2806', 'Lipid Metabolism Pathway WP3965', 'Target Of Rapamycin (TOR) Signaling WP1471', 'Caloric restriction and aging WP4191', 'Somatroph axis (GH) and its relationship to dietary restriction and aging WP4186', 'Association Between Physico-Chemical Features and Toxicity Associated Pathways WP3680', 'Breast cancer pathway WP4262', 'Pancreatic adenocarcinoma pathway WP4263', 'Non-small cell lung cancer WP4255', 'PDGFR-beta pathway WP3972', 'TGF-beta Receptor Signaling WP560', 'Pathways Affected in Adenoid Cystic Carcinoma WP3651', '4-hydroxytamoxifen, Dexamethasone, and Retinoic Acids Regulation of p27 Expression WP3879', 'Serotonin Receptor 4/6/7 and NR3C Signaling WP734', 'Prion disease pathway WP3995', 'Hedgehog Signaling Pathway WP4249', 'TYROBP Causal Network WP3945', 'Mitochondrial LC-Fatty Acid Beta-Oxidation WP368', 'Thermogenesis WP4321', 'Type II interferon signaling (IFNG) WP619', 'Hereditary leiomyomatosis and renal cell carcinoma pathway WP4206', 'Ferroptosis WP4313', 'Mevalonate pathway WP3963', 'Cholesterol Biosynthesis Pathway WP197', 'PI3K/AKT/mTOR - VitD3 Signalling WP4141', 'Factors and pathways affecting insulin-like growth factor (IGF1)-Akt signaling WP3850', 'Insulin signalling in human adipocytes (normal condition) WP3634', 'miRNA regulation of prostate cancer signaling pathways WP3981', 'Signaling Pathways in Glioblastoma WP2261', 'Endometrial cancer WP4155', 'MET in type 1 papillary renal cell carcinoma WP4205', 'Signaling of Hepatocyte Growth Factor Receptor WP313', 'PDGF Pathway WP2526', 'IL-2 Signaling Pathway WP49', 'IL-5 Signaling Pathway WP127', 'Oncostatin M Signaling Pathway WP2374', 'Interleukin-11 Signaling Pathway WP2332', 'IL-3 Signaling Pathway WP286', 'IL-7 Signaling Pathway WP205', 'IL-9 Signaling Pathway WP22', 'IL-10 Anti-inflammatory Signaling Pathway  WP4495', 'The human immune response to tuberculosis WP4197', 'Photodynamic therapy-induced HIF-1 survival signaling WP3614', 'Glutathione metabolism WP100', 'Gamma-Glutamyl Cycle for the biosynthesis and degradation of glutathione, including diseases WP4518', 'Thiamine metabolic pathways WP4297', 'Macrophage markers WP4146', 'Type II diabetes mellitus WP1584', 'Insulin signalling in human adipocytes (diabetic condition) WP3635', 'Alpha 6 Beta 4 signaling pathway WP244', 'DNA Damage Response (only ATM dependent) WP710', 'ErbB Signaling Pathway WP673', 'Integrin-mediated Cell Adhesion WP185', 'Focal Adhesion WP306', 'Kit receptor signaling pathway WP304', 'IL-4 Signaling Pathway WP395', 'Thymic Stromal LymphoPoietin (TSLP) Signaling Pathway WP2203', 'Prolactin Signaling Pathway WP2037', 'Leptin signaling pathway WP2034', 'IL-6 signaling pathway WP364', 'EPO Receptor Signaling WP581', 'TCA Cycle Nutrient Utilization and Invasiveness of Ovarian Cancer WP2868', 'Type III interferon signaling WP2113', 'GPCRs, Class C Metabotropic glutamate, pheromone WP501', 'Prostaglandin Synthesis and Regulation WP98', 'Eicosanoid Synthesis WP167', 'Purine metabolism WP4224', 'Pyrimidine metabolism and related diseases WP4225', 'Nucleotide Metabolism WP404', 'Methionine De Novo and Salvage Pathway WP3580', 'Methylation Pathways WP704', 'Hepatitis C and Hepatocellular Carcinoma WP3646', 'Chromosomal and microsatellite instability in colorectal cancer  WP4216', 'EGF/EGFR Signaling Pathway WP437', 'Insulin Signaling WP481', 'Regulation of Actin Cytoskeleton WP51', 'Leptin Insulin Overlap WP3935', 'Estrogen Receptor Pathway WP2881', 'Interferon type I signaling pathways WP585', 'Androgen receptor signaling pathway WP138', 'VEGFA-VEGFR2 Signaling Pathway WP3888', 'miRs in Muscle Cell Differentiation WP2012', 'Common Pathways Underlying Drug Addiction WP2636', 'Endothelin Pathways WP2197', 'Alanine and aspartate metabolism WP106', 'GABA receptor Signaling WP4159', 'HIF1A and PPARG regulation of glycolysis WP2456', 'Cori Cycle WP1946', 'Ganglio Sphingolipid Metabolism WP1423', 'One Carbon Metabolism WP241', 'Trans-sulfuration and one carbon metabolism WP2525', 'Trans-sulfuration pathway WP2333', 'Methionine metabolism leading to Sulphur Amino Acids and related disorders WP4292', 'One carbon metabolism and related pathways WP3940', 'Photodynamic therapy-induced NFE2L2 (NRF2) survival signaling WP3612', 'Physiological and Pathological Hypertrophy  of the Heart WP1528', 'MicroRNAs in cardiomyocyte hypertrophy WP1544', 'G13 Signaling Pathway WP524', 'Dual hijack model of Vif in HIV infection WP3300', 'Sulindac Metabolic Pathway WP2542', '', 'Hypothetical Craniofacial Development Pathway WP3655', 'Aflatoxin B1 metabolism WP699', 'G Protein Signaling Pathways WP35', 'T-Cell Receptor and Co-stimulatory Signaling WP2583', 'NAD metabolism, sirtuins and aging WP3630', 'NAD+ metabolism WP3644', 'Nuclear Receptors in Lipid Metabolism and Toxicity WP299', 'Drug Induction of Bile Acid Pathway WP2289', 'Glycolysis and Gluconeogenesis WP534', 'Globo Sphingolipid Metabolism WP1424', 'Disorders of Folate Metabolism and Transport WP4259', 'Folate-Alcohol and Cancer Pathway Hypotheses WP1589', 'Cysteine and methionine catabolism WP4504', 'Iron metabolism in placenta WP2007', 'NRF2-ARE regulation WP4357', 'Phytochemical activity on NRF2 transcriptional activation WP3', 'Initiation of transcription and translation elongation at the HIV-1 LTR WP3414', 'Cardiac Hypertrophic Response WP2795', 'Melatonin metabolism and effects WP3298', 'Vitamin D Metabolism WP1531', 'miRNA targets in ECM and membrane receptors WP2911', 'Simplified Interaction Map Between LOXL4 and Oxidative Stress Pathway WP3670', 'Nucleotide-binding Oligomerization Domain (NOD) pathway WP1433', 'Benzene metabolism WP3891', 'Heme Biosynthesis WP561', 'Notch Signaling Pathway WP61', 'Notch Signaling WP268', 'NAD+ biosynthetic pathways WP3645', 'Farnesoid X Receptor  Pathway WP2879', 'Pathways in clear cell renal cell carcinoma WP4018', 'Metabolic reprogramming in colon cancer WP4290', 'Pentose Phosphate Metabolism WP134', 'Ethanol effects on histone modifications WP3996', 'Vitamin B12 Disorders WP4271', 'Arrhythmogenic Right Ventricular Cardiomyopathy WP2118', 'Hippo-Merlin Signaling Dysregulation WP4541', 'Pathways Regulating Hippo Signaling WP4540', 'Osteoblast Signaling WP322', 'Osteoclast Signaling WP12', 'Neovascularisation processes WP4331', 'Differentiation of white and brown adipocyte    WP2895', 'Liver X Receptor Pathway WP2874', 'miR-509-3p alteration of YAP1/ECM axis WP3967', 'Overview of nanoparticle effects WP3287', 'Amyotrophic lateral sclerosis (ALS) WP2447', 'Vitamin D Receptor Pathway WP2877', '', 'NOTCH1 regulation of human endothelial cell calcification WP3413', 'Canonical  and Non-canonical Notch signaling WP3845', 'Oxysterols derived from cholesterol WP4545', 'Bile Acids synthesis and enterohepatic circulation  WP4389', 'Hypertrophy Model WP516', 'PPAR signaling pathway WP3942', 'PPAR Alpha Pathway WP2878', 'Valproic acid pathway WP3871', 'Pathogenic Escherichia coli infection WP2272', 'Parkin-Ubiquitin Proteasomal System pathway WP2359', 'Parkinsons Disease Pathway WP2371', 'Energy Metabolism WP1541', 'Mitochondrial Gene Expression WP391', 'Hematopoietic Stem Cell Gene Regulation by GABP alpha/beta Complex WP3657', 'Hypothesized Pathways in Pathogenesis of Cardiovascular Disease WP3668', 'TGF-B Signaling in Thyroid Cells for Epithelial-Mesenchymal Transition WP3859', 'SREBF and miR33 in cholesterol and lipid homeostasis WP2011', 'Sterol Regulatory Element-Binding Proteins (SREBP) signalling WP1982', 'Nanoparticle-mediated activation of receptor signaling WP2643', 'Photodynamic therapy-induced AP-1 survival signaling. WP3611', 'Endochondral Ossification WP474', 'Ebola Virus Pathway on Host WP4217', 'Chemokine signaling pathway WP3929', '', 'miR-517 relationship with ARCN1 and USP1 WP3596', 'Hfe effect on hepcidin production WP3924', 'Interleukin-1 Induced Activation of NF-kappa-B WP3656', 'Fluoropyrimidine Activity WP1601', 'Pregnane X Receptor pathway WP2876', 'Constitutive Androstane Receptor Pathway WP2875', 'Inflammatory Response Pathway WP453', 'Proteasome Degradation WP183', 'Dopamine metabolism WP2436', 'Biogenic Amine Synthesis WP550', 'Acetylcholine Synthesis WP528', 'MTHFR deficiency WP4288', 'Canonical and Non-Canonical TGF-B signaling WP3874', 'Bone Morphogenic Protein (BMP) Signalling and Regulation WP1425', 'BMP Signaling Pathway in Eyelid Development WP3927', 'MFAP5-mediated ovarian cancer cell motility and invasiveness WP3301', 'Inhibition of exosome biogenesis and secretion by Manumycin A in CRPC cells WP4301', 'MAPK Cascade WP422', 'Ras Signaling WP4223', 'TGF-beta Signaling Pathway WP366', 'Corticotropin-releasing hormone signaling pathway WP2355', 'Spinal Cord Injury WP2431', 'ID signaling pathway WP53', 'Simplified Depiction of MYD88 Distinct Input-Output Pathway WP3877', 'Fibrin Complement Receptor 3 Signaling Pathway WP4136', 'Pyrimidine metabolism WP4022', 'Nuclear Receptors Meta-Pathway WP2882', 'Aryl Hydrocarbon Receptor Pathway WP2873', 'Selective expression of chemokine receptors during T-cell polarization WP4494', 'Allograft Rejection WP2328', 'Matrix Metalloproteinases WP129', 'Thyroxine (Thyroid Hormone) Production WP1981', '', 'SCFA and skeletal muscle substrate metabolism WP4030', 'Glial Cell Differentiation WP2276', 'Oligodendrocyte Specification and differentiation(including remyelination), leading to Myelin Components for CNS WP4304', 'Heart Development WP1591', 'FTO Obesity Variant Mechanism WP3407', \"ERK Pathway in Huntington's Disease WP3853\", 'Serotonin Receptor 2 and ELK-SRF/GATA4 signaling WP732', 'Glycogen Synthesis and Degradation WP500', 'Oxidation by Cytochrome P450 WP43', 'Serotonin and anxiety-related events WP3944', 'Serotonin and anxiety WP3947', 'Non-genomic actions of 1,25 dihydroxyvitamin D3 WP4341', 'Regulation of toll-like receptor signaling pathway WP1449', 'Toll-like Receptor Signaling WP3858', 'Eukaryotic Transcription Initiation WP405', 'NRF2 pathway WP2884', 'Lung fibrosis WP3624', 'Cytokines and Inflammatory Response WP530', 'Development and heterogeneity of the ILC family WP3893', 'Glucuronidation WP698', 'Vitamin B6-dependent and responsive disorders WP4228', 'mRNA Processing WP411', 'Mammary gland development pathway - Pregnancy and lactation (Stage 3 of 4) WP2817', 'Mammary gland development pathway - Puberty (Stage 2 of 4) WP2814', 'Urea cycle and metabolism of amino groups WP497', 'Molybdenum cofactor (Moco) biosynthesis WP4507', 'Effects of Nitric Oxide WP1995', 'Monoamine GPCRs WP58', 'Sleep regulation WP3591', 'Nucleotide GPCRs WP80', 'Metapathway biotransformation Phase I and II WP702', 'Arachidonate Epoxygenase / Epoxide Hydrolase WP678', 'Translation Factors WP107', 'Synaptic signaling pathways associated with autism spectrum disorder WP4539', 'Toll-like Receptor Signaling Pathway WP75', 'Angiopoietin Like Protein 8 Regulatory Pathway WP3915', 'T-Cell antigen Receptor (TCR)  Signaling Pathway WP69', 'T-Cell antigen Receptor (TCR) pathway during Staphylococcus aureus infection WP3863', 'Regulation of Microtubule Cytoskeleton WP2038', 'MicroRNA for Targeting Cancer Growth and Vascularization in Glioblastoma WP3593', '', 'Synaptic Vesicle Pathway WP2267', 'Cerebral Organic Acidurias, including diseases WP4519', 'Nuclear Receptors WP170', 'Ovarian Infertility Genes WP34', 'Platelet-mediated interactions with vascular and circulating cells WP4462', 'Amino Acid metabolism WP3925', 'Splicing factor NOVA regulated synaptic proteins WP4148', 'NAD Biosynthesis II (from tryptophan) WP2485', 'Tryptophan catabolism leading to NAD+ production WP4210', '', 'Metabolism of Spingolipids in ER and Golgi apparatus WP4142', 'Sulfation Biotransformation Reaction WP692', 'Intraflagellar transport proteins binding to dynein WP4532', 'Microglia Pathogen Phagocytosis Pathway WP3937', 'PI3K-AKT-mTOR signaling pathway and therapeutic opportunities WP3844', 'BDNF-TrkB Signaling WP3676', 'RAC1/PAK1/p38/MMP2 Pathway WP3303', 'B Cell Receptor Signaling Pathway WP23', 'Brain-Derived Neurotrophic Factor (BDNF) signaling pathway WP2380', 'AGE/RAGE pathway WP2324', 'Angiogenesis WP1539', 'Robo4 and VEGF Signaling Pathways Crosstalk WP3943', 'Serotonin Transporter Activity WP1455', 'Adipogenesis WP236', 'White fat cell differentiation WP4149', 'Preimplantation Embryo WP3527', 'Extracellular vesicle-mediated signaling in recipient cells WP2870', 'Cytoplasmic Ribosomal Proteins WP477', 'Degradation pathway of sphingolipids, including diseases WP4153', 'Tryptophan metabolism WP465', '', 'Secretion of Hydrochloric Acid in Parietal Cells WP2597', 'Transcription co-factors SKI and SKIL protein partners WP4533', 'Rett syndrome causing genes WP4312', 'Genes related to primary cilium development (based on CRISPR) WP4536', 'Ciliary landscape WP4352', 'RalA downstream regulated genes WP2290', 'TNF alpha Signaling Pathway WP231', 'Oxidative Damage WP3941', 'Aryl Hydrocarbon Receptor WP2586', 'Human Thyroid Stimulating Hormone (TSH) signaling pathway WP2032', 'Follicle Stimulating Hormone (FSH) signaling pathway WP2035', 'Type 2 papillary renal cell carcinoma WP4241', 'Amplification and Expansion of Oncogenic Pathways as Metastatic Traits WP3678', 'Monoamine Transport WP727', 'Transcription factor regulation in adipogenesis WP3599', 'Transcriptional cascade regulating adipogenesis WP4211', 'EV release from cardiac cells and their functional effects WP3297', 'Regulation of Wnt/B-catenin Signaling by Small Molecule Compounds WP3664', 'Hair Follicle Development: Cytodifferentiation (Part 3 of 3) WP2840', 'Ectoderm Differentiation WP2858', 'Dopaminergic Neurogenesis WP2855', 'Nicotine Activity on Dopaminergic Neurons WP1602', 'Cell-type Dependent Selectivity of CCK2R Signaling WP3679', 'Gastric acid production WP2596', 'Cytosine methylation WP3585', 'TCA Cycle (aka Krebs or citric acid cycle) WP78', 'TCA Cycle and Deficiency of Pyruvate Dehydrogenase complex (PDHc) WP2453', 'p38 MAPK Signaling Pathway WP400', 'Apoptosis WP254', 'miRNAs involvement in the immune response in sepsis WP4329', 'IL-1 signaling pathway WP195', 'Bladder Cancer WP2828', 'miRNAs involved in DNA damage response WP1545', 'LncRNA-mediated mechanisms of therapeutic resistance WP3672', 'Vitamin A and Carotenoid Metabolism WP716', 'Triacylglyceride Synthesis WP325', 'Sphingolipid pathway WP1422', 'mir-124 predicted interactions with cell cycle and differentiation  WP3595', 'Striated Muscle Contraction Pathway WP383', 'Wnt Signaling Pathway WP363', 'Wnt/beta-catenin Signaling Pathway in Leukemia WP3658', 'Neural Crest Differentiation WP2064', 'Hippo-Yap signaling pathway WP4537', 'Glucocorticoid and Mineralcorticoid Metabolism WP237', 'Classical pathway of steroidogenesis, including diseases WP4523', 'Deregulation of Rab and Rab Effector Genes in Bladder Cancer WP2291', 'Gastric Cancer Network 2 WP2363', 'Disorders of the Krebs cycle WP4236', '', 'Nanoparticle triggered regulated necrosis WP2513', 'RIG-I-like Receptor Signaling WP3865'], [0.00236, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00651, 0.0, 0.00234, 0.00447, 0.0, 0.0, 0.0, 0.0026, 0.00259, 0.00245, 0.0, 0.0, 0.0, 0.0, 0.00248, 0.00644, 0.0, 0.00251, 0.00251, 0.0, 0.0, 0.00471, 0.00227, 0.00238, 0.0, 0.00218, 0.00215, 0.0, 0.0, 0.00685, 0.0019, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00249, 0.0, 0.00521, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00258, 0.0, 0.0, 0.00252, 0.00231, 0.00563, 0.0, 0.0, 0.0, 0.00504, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00253, 0.0, 0.0, 0.00434, 0.0, 0.0, 0.00628, 0.0046, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00254, 0.0, 0.0, 0.00241, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00246, 0.0, 0.00243, 0.00408, 0.00377, 0.00699, 0.00377, 0.00696, 0.0, 0.0, 0.0, 0.0, 0.00234, 0.0, 0.0, 0.0, 0.00215, 0.00233, 0.0, 0.00383, 0.00194, 0.00226, 0.0, 0.00611, 0.00426, 0.0021, 0.00382, 0.00737, 0.00676, 0.0026, 0.0, 0.0, 0.00243, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00256, 0.00248, 0.00244, 0.0, 0.00562, 0.0, 0.0, 0.00424, 0.00248, 0.00243, 0.0, 0.0, 0.00227, 0.00378, 0.00216, 0.0, 0.0, 0.0, 0.00227, 0.0, 0.00508, 0.0, 0.00477, 0.0, 0.0051, 0.00621, 0.0, 0.0, 0.00241, 0.0, 0.0, 0.0, 0.00246, 0.00261, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0024, 0.00241, 0.00227, 0.00239, 0.00236, 0.0, 0.0, 0.0, 0.00251, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00261, 0.0, 0.00412, 0.00215, 0.0021, 0.00524, 0.0023, 0.00233, 0.0, 0.00222, 0.00222, 0.00239, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00252, 0.00521, 0.0, 0.0, 0.00559, 0.00748, 0.0019, 0.0, 0.0, 0.00466, 0.00215, 0.00655, 0.00248, 0.0024, 0.00245, 0.0, 0.00246, 0.0, 0.0, 0.00258, 0.00494, 0.00246, 0.0026, 0.00777, 0.00468, 0.0, 0.0025, 0.0, 0.00242, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00214, 0.0, 0.00259, 0.0, 0.0, 0.0, 0.00238, 0.0, 0.0, 0.0026, 0.0, 0.00258, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0026, 0.0, 0.0, 0.0, 0.0, 0.00652, 0.0024, 0.0, 0.00246, 0.00515, 0.00223, 0.00606, 0.00634, 0.0, 0.0, 0.00253, 0.0, 0.0, 0.0, 0.0, 0.00242, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00258, 0.0, 0.00452, 0.00499, 0.0, 0.0, 0.00225, 0.00242, 0.00474, 0.00761, 0.0, 0.0, 0.0, 0.0, 0.0045, 0.00248, 0.0, 0.00228, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00458, 0.00258, 0.0, 0.0, 0.0025, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00537, 0.00197, 0.00214, 0.0, 0.0, 0.0, 0.0, 0.00217, 0.00576, 0.0, 0.0, 0.00216, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00257, 0.0, 0.00482, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00249, 0.0, 0.00399, 0.00246, 0.0, 0.0, 0.00262, 0.0, 0.0, 0.0, 0.0, 0.00717, 0.0, 0.00234, 0.00235, 0.0, 0.00394, 0.0, 0.0, 0.00238, 0.0, 0.0, 0.0, 0.00262, 0.00242, 0.0, 0.0, 0.01073, 0.00719, 0.00522, 0.00512, 0.0, 0.0, 0.0, 0.00498, 0.0, 0.0, 0.00244, 0.0, 0.0, 0.00385, 0.0, 0.00251, 0.00262, 0.0, 0.00594, 0.00246, 0.0, 0.0, 0.00216, 0.0, 0.00475, 0.0, 0.0, 0.0, 0.0, 0.00418, 0.00846, 0.0, 0.00214, 0.0, 0.0, 0.00227, 0.00498, 0.0, 0.0, 0.0, 0.00252, 0.0, 0.0, 0.0, 0.00216, 0.00195, 0.0, 0.00253, 0.0, 0.0, 0.0, 0.0, 0.00512, 0.00244, 0.00218, 0.0, 0.0, 0.00241, 0.0, 0.0, 0.00239, 0.0, 0.00249, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00261, 0.00256, 0.0, 0.0, 0.00262, 0.0, 0.0, 0.0023])\n})\n</script>\n"
     },
     "metadata": {}
    },
    {
     "output_type": "stream",
     "name": "stdout",
     "text": "The color scaling factor for this canvas is: 0.01073\nGene sets with a similarity index less than 0.00611 are not colored\n"
    }
   ],
   "source": [
    "if len(all_libraries) == 1:\n",
    "    anneal_list, x_dimension, y_dimension = library_processing(0)\n",
    "    color_list, scaling_factor, cut_off_value = get_color(anneal_list, 2.0, scaling_factor, x_dimension, y_dimension)\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "    Canvas(x_dimension, y_dimension, color_list, list(unzipped_anneal_list[0]), list(unzipped_anneal_list[2]))\n",
    "    print(\"The color scaling factor for this canvas is: \" + str(scaling_factor))\n",
    "    print(\"Gene sets with a similarity index less than \" + str(cut_off_value) + \" are not colored\")\n",
    "\n",
    "elif len(all_libraries) == 2:\n",
    "    # first library\n",
    "    anneal_list_1, x_dimension_1, y_dimension_1 = library_processing(0)\n",
    "    color_list_1, scaling_factor_1, cut_off_value_1 = get_color(anneal_list_1, 2.0, scaling_factor, x_dimension_1, y_dimension_1)\n",
    "\n",
    "    # second library\n",
    "    anneal_list_2, x_dimension_2, y_dimension_2 = library_processing(1)\n",
    "    color_list_2, scaling_factor_2, cut_off_value_2 = get_color(anneal_list_2, 2.0, scaling_factor, x_dimension_2, y_dimension_2)\n",
    "\n",
    "    if scaling_factor_1 > scaling_factor_2:\n",
    "        scaling_factor = scaling_factor_1\n",
    "        color_list_2, scaling_factor_2, cut_off_value_2 = get_color(anneal_list_2, 2.0, scaling_factor, x_dimension_2, y_dimension_2)\n",
    "    else:\n",
    "        scaling_factor = scaling_factor_2\n",
    "        color_list_1, scaling_factor_1, cut_off_value_1 = get_color(anneal_list_1, 2.0, scaling_factor, x_dimension_1, y_dimension_1)\n",
    "\n",
    "    unzipped_anneal_list_1 = unzip_list(anneal_list_1)\n",
    "    unzipped_anneal_list_2 = unzip_list(anneal_list_2)\n",
    "\n",
    "    Canvas(x_dimension_1, y_dimension_1, color_list_1, list(unzipped_anneal_list_1[0]), list(unzipped_anneal_list_1[2]))\n",
    "    Canvas(x_dimension_2, y_dimension_2, color_list_2, list(unzipped_anneal_list_2[0]), list(unzipped_anneal_list_2[2]))\n",
    "\n",
    "    print(\"The color scaling factor for these canvases are: \" + str(scaling_factor))\n",
    "    print(\"For the first canvas, gene sets with a similarity index less than \" + str(cut_off_value_1) + \" are not colored\")\n",
    "    print(\"For the second canvas, gene sets with a similarity index less than \" + str(cut_off_value_2) + \" are not colored\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 64-bit ('py383': conda)",
   "language": "python",
   "name": "python38364bitpy383conda2309aa976eb546f3b7675a3d2adb70e0"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}