{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Canvas Creation from Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import json\n",
    "import requests\n",
    "import math\n",
    "import matplotlib\n",
    "import uuid\n",
    "import urllib\n",
    "from textwrap import dedent\n",
    "from IPython.core.display import display, HTML\n",
    "from string import Template\n",
    "from random import seed, randint\n",
    "from operator import itemgetter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# either name(s) (up to 2) of Enrichr library or name(s) of local library file\n",
    "all_libraries = ['WikiPathways_2019_Human'] \n",
    "\n",
    "color = 'Purple'\n",
    "scaling_factor = ''\n",
    "num_hex_colored = '10'\n",
    "annealing_steps = '750' # probably want this to be around 50k - 100k for the best grouping\n",
    "\n",
    "# list of genes that was input\n",
    "# assumes gene list is stored in geneList.txt\n",
    "open_gene_list_file = open('geneList.txt','r')\n",
    "lines = open_gene_list_file.readlines()\n",
    "genes = [x.strip().upper() for x in lines]\n",
    "open_gene_list_file.close()\n",
    "\n",
    "#genes = ['TP53', 'TNF', 'EGFR', 'GKN1', 'HADHA', 'APOE', 'ESR1', 'VEGFA', 'TGFB1', 'PREPL', 'TIA1', 'TPO', 'TTN', 'SATB2', 'CHPF', 'MALL', 'MIPIP', 'NUPL1', 'IL6', 'PDIA3', 'CTNNB1', 'SLC39A1', 'DTNA','SLC1A1', 'GALNT2', 'HIST2H2AC', 'CD63']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def library_processing(library_index):\n",
    "    # processes library data\n",
    "    raw_library_data = []\n",
    "    library_data = []\n",
    "\n",
    "    # open Enrichr library from online\n",
    "    with urllib.request.urlopen('https://amp.pharm.mssm.edu/Enrichr/geneSetLibrary?mode=text&libraryName=' + all_libraries[library_index]) as f:\n",
    "        for line in f.readlines():\n",
    "                raw_library_data.append(line.decode(\"utf-8\").split(\"\\t\\t\"))\n",
    "\n",
    "    # OR\n",
    "    # locally upload library\n",
    "    # with open('filepath', 'r') as f:\n",
    "    #     for line in f.readlines():\n",
    "    #        raw_library_data.append(line.split(\"\\t\\t\"))\n",
    "\n",
    "    name = []\n",
    "    gene_list = []\n",
    "\n",
    "    for i in range(len(raw_library_data)):\n",
    "        name += [raw_library_data[i][0]]\n",
    "        raw_genes = raw_library_data[i][1].split(\"\\t\")\n",
    "        gene_list += [raw_genes[:-1]]\n",
    "\n",
    "    # library_data: a 2D list where the first element is the name and the second element is a list of genes \n",
    "    # associated with that name\n",
    "    library_data = [list(a) for a in zip(name, gene_list)]\n",
    "\n",
    "    jaccard_indices = []\n",
    "    indices = []\n",
    "\n",
    "    for gene_set in library_data:\n",
    "        intersection = [value for value in gene_set[1] if value in genes]\n",
    "        index = len(intersection)/(len(gene_set[1]) + len(genes))\n",
    "        jaccard_indices += [[gene_set[0], index]]\n",
    "        indices += [round(index, 5)]\n",
    "\n",
    "    # determine the dimensions of the canvas\n",
    "    x_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "    y_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "\n",
    "    # zip name, gene_list, indices, and blank list for neighbor score then add dummy entries to the zipped list\n",
    "    num_hex = x_dimension*y_dimension\n",
    "    neighbor_score = [0.0] * len(name)\n",
    "    anneal_list = list(zip(name, gene_list, indices, neighbor_score))\n",
    "\n",
    "    unzipped_anneal_list = [name, gene_list, indices, neighbor_score]\n",
    "\n",
    "    # add \"dummy\" hexagons so the rectangular shape is filled\n",
    "    for i in range(len(name), num_hex):\n",
    "        anneal_list += [('', [], 0.0, 0.0)]\n",
    "\n",
    "    return anneal_list, x_dimension, y_dimension\n",
    "\n",
    "def unzip_list(anneal_list):\n",
    "    unzipped_list = zip(*anneal_list)\n",
    "    return list(unzipped_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_color(anneal_list, cut_off_value, scaling_factor, x_dimension, y_dimension):\n",
    "\n",
    "    # Deal with cut_off_value (only color the most significant 10/20 hexagons)\n",
    "    if cut_off_value == 2.0:\n",
    "        sort_list = sorted(anneal_list, key=itemgetter(2), reverse=True)\n",
    "        cut_off_value = sort_list[int(num_hex_colored)-1][2]\n",
    "\n",
    "    r_value = 0\n",
    "    g_value = 0\n",
    "    b_value = 0\n",
    "\n",
    "    if color == 'Red':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.8\n",
    "        b_value = 0.8\n",
    "    if color == 'Yellow':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.3\n",
    "        b_value = 1.0\n",
    "    if color == 'Purple':\n",
    "        r_value = 0.5\n",
    "        g_value = 1.0\n",
    "        b_value = 0.0\n",
    "    if color == 'Pink':\n",
    "        r_value = 0.0\n",
    "        g_value = 1.0\n",
    "        b_value = 0.2\n",
    "    if color == 'Orange':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.45\n",
    "        b_value = 1.0\n",
    "    if color == 'Green':\n",
    "        r_value = 1.0\n",
    "        g_value = 0.0\n",
    "        b_value = 1.0\n",
    "    if color == 'Blue':\n",
    "        r_value = 1.0\n",
    "        g_value = 0.9\n",
    "        b_value = 0.0\n",
    "\n",
    "    color_list = []\n",
    "\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "    \n",
    "    if scaling_factor == '':\n",
    "        max_index = max(unzipped_anneal_list[2])\n",
    "    else:\n",
    "        max_index = float(scaling_factor)\n",
    "\n",
    "    scaled_list = [i/max_index for i in unzipped_anneal_list[2]]\n",
    "\n",
    "    for i in range(x_dimension*y_dimension):\n",
    "        if i < len(unzipped_anneal_list[2]) and float(unzipped_anneal_list[2][i]) >= cut_off_value:\n",
    "            color_list += [matplotlib.colors.to_hex((1-scaled_list[i]*r_value, \n",
    "            1-scaled_list[i]*g_value, 1-scaled_list[i]*b_value))]\n",
    "        elif i < len(unzipped_anneal_list[2]):\n",
    "            color_list += [matplotlib.colors.to_hex((1-scaled_list[i], \n",
    "            1-scaled_list[i], 1-scaled_list[i]))]\n",
    "        else:\n",
    "            color_list += [\"#FFFFFF\"]\n",
    "    return color_list, max_index, cut_off_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_neighbors(ind, x_dimension, y_dimension):\n",
    "    # returns a list of the indices of the neighbors of the index given\n",
    "    \n",
    "    neighbors = []\n",
    "    num_hex = x_dimension * y_dimension\n",
    "\n",
    "    if 0 <= ind <= x_dimension-1:\n",
    "        # top row (inc. corners)\n",
    "        if ind == 0:\n",
    "            # top left corner\n",
    "            neighbors = [num_hex-1, num_hex-x_dimension, x_dimension-1, 2*x_dimension-1, ind+1, ind+x_dimension]\n",
    "        elif ind == (x_dimension-1):\n",
    "            # top right corner\n",
    "            neighbors = [ind-1, ind+x_dimension, 0, ind+x_dimension-1, num_hex-2, num_hex-1]\n",
    "        else:\n",
    "            # non-corner top row\n",
    "            neighbors = [ind-1, ind+1, ind+x_dimension, ind+num_hex-x_dimension-1, \n",
    "            ind+num_hex-x_dimension, ind+x_dimension-1]\n",
    "\n",
    "    elif (num_hex - x_dimension) <= ind <= num_hex -1:\n",
    "        if ind == (num_hex-x_dimension):\n",
    "            # bottom left corner\n",
    "            neighbors = [ind+1, ind-x_dimension, ind-x_dimension+1, 0, 1, num_hex-1]\n",
    "        elif ind == (num_hex-1):\n",
    "            # bottom right corner\n",
    "            neighbors = [ind-1, ind-x_dimension, ind-x_dimension+1, 0, x_dimension-1,\n",
    "            num_hex-2*x_dimension]\n",
    "        else:\n",
    "            # non-corner bottom row\n",
    "            neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind-num_hex+x_dimension,\n",
    "            ind-num_hex+x_dimension+1]\n",
    "    elif ind % y_dimension == 0 and (ind/y_dimension)%2 == 1:\n",
    "        # \"inner\" left edge (not top or bottom row)\n",
    "        neighbors = [ind+x_dimension-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension+1]\n",
    "    elif ind % y_dimension == 0 and (ind/y_dimension)%2 == 0:\n",
    "        # \"outer\" left edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind+1, ind+x_dimension, ind+2*x_dimension-1, ind-x_dimension, \n",
    "        ind+x_dimension-1]\n",
    "    elif (ind+1) % y_dimension == 0 and ((ind+1)/y_dimension)%2 == 0:\n",
    "        # \"outer\" right edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind-2*x_dimension+1]\n",
    "    elif (ind+1) % y_dimension == 0 and ((ind+1)/y_dimension)%2 == 1:\n",
    "        # \"inner\" right edge (not top or bottom row)\n",
    "        neighbors = [ind-1, ind-x_dimension-1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension-1]\n",
    "    else:\n",
    "        # middle\n",
    "        neighbors = [ind-1, ind+1, ind-x_dimension, ind-x_dimension+1, ind+x_dimension, \n",
    "        ind+x_dimension+1]\n",
    "    return neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initially find fitness\n",
    "def find_fitness(anneal_list, x_dimension, y_dimension):\n",
    "    fitness = 0\n",
    "    for i in range(len(anneal_list)):\n",
    "        neighbors = find_neighbors(i, x_dimension, y_dimension)\n",
    "        sum_neighbor_score = 0\n",
    "        for index in neighbors:\n",
    "            intersection = [value for value in anneal_list[index][1] if value in anneal_list[i][1]]\n",
    "            if len(anneal_list[index][1]) + len(anneal_list[i][1]) != 0:\n",
    "                jaccard = len(intersection)/(len(anneal_list[index][1]) + len(anneal_list[i][1]))\n",
    "            else:\n",
    "                jaccard = 0.0\n",
    "            sum_neighbor_score += jaccard\n",
    "        hex_list = list(anneal_list[i])\n",
    "        hex_list[3] = sum_neighbor_score\n",
    "        hex_tuple = tuple(hex_list)\n",
    "        anneal_list[i] = hex_tuple\n",
    "        fitness += sum_neighbor_score\n",
    "    return fitness, anneal_list\n",
    "\n",
    "# take indices of swapped hexagons\n",
    "def find_swapped_fitness(anneal_list, swapped_a, swapped_b, old_fitness, x_dimension, y_dimension):\n",
    "    neighbors_a = find_neighbors(swapped_a, x_dimension, y_dimension)\n",
    "    neighbors_b = find_neighbors(swapped_b, x_dimension, y_dimension)\n",
    "    hexagons_to_update = [swapped_a, swapped_b] + neighbors_a + neighbors_b\n",
    "    anneal_copy = anneal_list.copy()\n",
    "\n",
    "    new_fitness = 0\n",
    "    # Recalculate scores for all hexagons that need updating\n",
    "    for hex in hexagons_to_update:\n",
    "\n",
    "        # subtract out the swapped neighbor fitnesses because they are changing \n",
    "        old_fitness -= anneal_copy[hex][3]\n",
    "\n",
    "        neighbors = find_neighbors(hex, x_dimension, y_dimension)\n",
    "        sum_neighbor_score = 0\n",
    "        for index in neighbors:\n",
    "            intersection = [value for value in anneal_copy[index][1] if value in anneal_copy[hex][1]]\n",
    "            if len(anneal_copy[index][1]) + len(anneal_copy[hex][1]) != 0:\n",
    "                jaccard = len(intersection)/(len(anneal_copy[index][1]) + len(anneal_copy[hex][1]))\n",
    "            else:\n",
    "                jaccard = 0.0\n",
    "            sum_neighbor_score += jaccard\n",
    "        hex_list = list(anneal_copy[hex])\n",
    "        hex_list[3] = sum_neighbor_score\n",
    "        hex_tuple = tuple(hex_list)\n",
    "        anneal_copy[hex] = hex_tuple\n",
    "        new_fitness += sum_neighbor_score\n",
    "        \n",
    "    return old_fitness + new_fitness, anneal_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annealing(anneal_list, steps, old_fitness, x_dimension, y_dimension):\n",
    "    num_hex = x_dimension * y_dimension\n",
    "    # returns unzipped list\n",
    "    for i in range(steps):\n",
    "        index_a = randint(0, num_hex-1)\n",
    "        index_b = randint(0, num_hex-1)\n",
    "        anneal_list[index_a], anneal_list[index_b] = anneal_list[index_b], anneal_list[index_a]\n",
    "        new_fitness, new_anneal_list = find_swapped_fitness(anneal_list, index_a, index_b, old_fitness, x_dimension, y_dimension)\n",
    "\n",
    "        if new_fitness <= old_fitness:\n",
    "            # swap back\n",
    "            anneal_list[index_a], anneal_list[index_b] = anneal_list[index_b], anneal_list[index_a]\n",
    "        else:\n",
    "            # finalize the swap by resetting old_fitness and changing anneal_list\n",
    "            old_fitness = new_fitness\n",
    "            anneal_list = new_anneal_list\n",
    "    return anneal_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def init_chart():\n",
    "  chart_id = 'mychart-' + str(uuid.uuid4())\n",
    "  display(HTML('<script src=\"/static/components/requirejs/require.js\"></script>'))\n",
    "  display(HTML(Template(dedent('''\n",
    "  <script>\n",
    "  require.config({\n",
    "    paths: {\n",
    "      'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n",
    "      'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n",
    "    },\n",
    "    shim: {\n",
    "      'd3-hexbin': ['d3']\n",
    "    }\n",
    "  })\n",
    "\n",
    "  // If we configure mychart via url, we can eliminate this define here\n",
    "  define($chart_id, ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n",
    "    return function (figure_id, numA, numB, colorList, libraryList, indices) {\n",
    "      var margin = {top: 50, right: 20, bottom: 20, left: 50},\n",
    "        width = 850 - margin.left - margin.right,\n",
    "        height = 350 - margin.top - margin.bottom;\n",
    "\n",
    "      // append the svg object to the body of the page\n",
    "      var svG = d3.select('#' + figure_id)\n",
    "                  .attr(\"width\", width + margin.left + margin.right)\n",
    "                  .attr(\"height\", height + margin.top + margin.bottom)\n",
    "                  .append(\"g\")\n",
    "                  .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
    "      \n",
    "      //The number of columns and rows of the heatmap\n",
    "      var MapColumns = numA,\n",
    "          MapRows = numB;\n",
    "\n",
    "      //The maximum radius the hexagons can have to still fit the screen\n",
    "      var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n",
    "\n",
    "      //Calculate the center position of each hexagon\n",
    "      var points = [];\n",
    "      for (var i = 0; i < MapRows; i++) {\n",
    "          for (var j = 0; j < MapColumns; j++) {\n",
    "              var x = hexRadius * j * Math.sqrt(3)\n",
    "              //Offset each uneven row by half of a \"hex-width\" to the right\n",
    "              if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n",
    "              var y = hexRadius * i * 1.5\n",
    "              points.push([x,y])\n",
    "          }\n",
    "      }\n",
    "\n",
    "      //Set the hexagon radius\n",
    "      var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n",
    "\n",
    "      svG.append(\"g\")\n",
    "        .selectAll(\".hexagon\")\n",
    "        .data(hexbin(points))\n",
    "        .enter().append(\"path\")\n",
    "        .attr(\"class\", \"hexagon\")\n",
    "        .attr(\"d\", function (d) {\n",
    "            return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n",
    "        })\n",
    "        .attr(\"stroke\", \"black\")\n",
    "        .attr(\"stroke-width\", \"1px\")\n",
    "        .style(\"fill\", function (d,i) { return colorList[i]; })\n",
    "        .on(\"mouseover\", mover)\n",
    "        .on(\"mouseout\", mout)\n",
    "        .append(\"svg:title\")\n",
    "        .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n",
    "\n",
    "      // Mouseover function\n",
    "      function mover(d) {\n",
    "      d3.select(this)\n",
    "        .transition().duration(10)  \n",
    "        .style(\"fill-opacity\", 0.3)\n",
    "      };\n",
    "\n",
    "      // Mouseout function\n",
    "      function mout(d) { \n",
    "      d3.select(this)\n",
    "        .transition().duration(10)\n",
    "        .style(\"fill-opacity\", 1)\n",
    "      };\n",
    "\n",
    "  }\n",
    "\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({ 'chart_id': repr(chart_id) })))\n",
    "  return chart_id\n",
    "\n",
    "def Canvas(numA, numB, colorList, libraryList, indices):\n",
    "  chart_id = init_chart()\n",
    "  display(HTML(Template(dedent('''\n",
    "  <svg id=$figure_id></svg>\n",
    "  <script>\n",
    "  require([$chart_id], function(mychart) {\n",
    "    mychart($figure_id, $numA, $numB, $colorList, $libraryList, $indices)\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({\n",
    "      'chart_id': repr(chart_id),\n",
    "      'figure_id': repr('fig-' + str(uuid.uuid4())),\n",
    "      'numA': repr(numA),\n",
    "      'numB': repr(numB),\n",
    "      'colorList': repr(colorList),\n",
    "      'libraryList': repr(libraryList),\n",
    "      'indices': repr(indices)\n",
    "  })))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script src=\"/static/components/requirejs/require.js\"></script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<script>\n",
       "require.config({\n",
       "  paths: {\n",
       "    'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n",
       "    'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n",
       "  },\n",
       "  shim: {\n",
       "    'd3-hexbin': ['d3']\n",
       "  }\n",
       "})\n",
       "\n",
       "// If we configure mychart via url, we can eliminate this define here\n",
       "define('mychart-9e07b6f5-2b55-47c9-bb37-a40cd0293e4d', ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n",
       "  return function (figure_id, numA, numB, colorList, libraryList, indices) {\n",
       "    var margin = {top: 50, right: 20, bottom: 20, left: 50},\n",
       "      width = 850 - margin.left - margin.right,\n",
       "      height = 350 - margin.top - margin.bottom;\n",
       "\n",
       "    // append the svg object to the body of the page\n",
       "    var svG = d3.select('#' + figure_id)\n",
       "                .attr(\"width\", width + margin.left + margin.right)\n",
       "                .attr(\"height\", height + margin.top + margin.bottom)\n",
       "                .append(\"g\")\n",
       "                .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
       "\n",
       "    //The number of columns and rows of the heatmap\n",
       "    var MapColumns = numA,\n",
       "        MapRows = numB;\n",
       "\n",
       "    //The maximum radius the hexagons can have to still fit the screen\n",
       "    var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n",
       "\n",
       "    //Calculate the center position of each hexagon\n",
       "    var points = [];\n",
       "    for (var i = 0; i < MapRows; i++) {\n",
       "        for (var j = 0; j < MapColumns; j++) {\n",
       "            var x = hexRadius * j * Math.sqrt(3)\n",
       "            //Offset each uneven row by half of a \"hex-width\" to the right\n",
       "            if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n",
       "            var y = hexRadius * i * 1.5\n",
       "            points.push([x,y])\n",
       "        }\n",
       "    }\n",
       "\n",
       "    //Set the hexagon radius\n",
       "    var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n",
       "\n",
       "    svG.append(\"g\")\n",
       "      .selectAll(\".hexagon\")\n",
       "      .data(hexbin(points))\n",
       "      .enter().append(\"path\")\n",
       "      .attr(\"class\", \"hexagon\")\n",
       "      .attr(\"d\", function (d) {\n",
       "          return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n",
       "      })\n",
       "      .attr(\"stroke\", \"black\")\n",
       "      .attr(\"stroke-width\", \"1px\")\n",
       "      .style(\"fill\", function (d,i) { return colorList[i]; })\n",
       "      .on(\"mouseover\", mover)\n",
       "      .on(\"mouseout\", mout)\n",
       "      .append(\"svg:title\")\n",
       "      .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n",
       "\n",
       "    // Mouseover function\n",
       "    function mover(d) {\n",
       "    d3.select(this)\n",
       "      .transition().duration(10)  \n",
       "      .style(\"fill-opacity\", 0.3)\n",
       "    };\n",
       "\n",
       "    // Mouseout function\n",
       "    function mout(d) { \n",
       "    d3.select(this)\n",
       "      .transition().duration(10)\n",
       "      .style(\"fill-opacity\", 1)\n",
       "    };\n",
       "\n",
       "}\n",
       "\n",
       "})\n",
       "</script>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<svg id='fig-2f2d83f9-9fbc-4e71-b6f2-7ad98f7222c6'></svg>\n",
       "<script>\n",
       "require(['mychart-9e07b6f5-2b55-47c9-bb37-a40cd0293e4d'], function(mychart) {\n",
       "  mychart('fig-2f2d83f9-9fbc-4e71-b6f2-7ad98f7222c6', 22, 22, ['#ffffff', '#ffffff', '#ffffff', '#c4c4c4', '#ffffff', '#c6c6c6', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#cccccc', '#858585', '#ffffff', '#ffffff', '#c7c7c7', '#cccccc', '#ffffff', '#cbcbcb', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#a5a5a5', '#ffffff', '#ffffff', '#ffffff', '#c2c2c2', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#aa55ff', '#c2c2c2', '#c1c1c1', '#c6c6c6', '#c3c3c3', '#c8c8c8', '#ffffff', '#c7c7c7', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#929292', '#ffffff', '#ffffff', '#9d9d9d', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#d1d1d1', '#ffffff', '#ffffff', '#c6c6c6', '#ffffff', '#c1c1c1', '#c2c2c2', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#797979', '#6e6e6e', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#8000ff', '#6a6a6a', '#c9c9c9', '#d1d1d1', '#ffffff', '#c5c5c5', '#8e8e8e', '#c2c2c2', '#6b6b6b', '#c5c5c5', '#ffffff', '#c3c3c3', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#aa54ff', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#c5c5c5', '#ffffff', '#c9c9c9', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#828282', '#c1c1c1', '#ffffff', '#898989', '#ffffff', '#ffffff', '#c5c5c5', '#ffffff', '#c5c5c5', '#ffffff', '#cccccc', '#ffffff', '#cacaca', '#a1a1a1', '#c2c2c2', '#c4c4c4', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#5c5c5c', '#ffffff', '#8a8a8a', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#8c8c8c', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#8e8e8e', '#c2c2c2', '#ffffff', '#c8c8c8', '#ffffff', '#727272', '#c1c1c1', '#ffffff', '#c4c4c4', '#6f6f6f', '#ffffff', '#c9c9c9', '#ffffff', '#9a9a9a', '#c9c9c9', '#ffffff', '#ffffff', '#c4c4c4', '#ffffff', '#ffffff', '#c6c6c6', '#c7c7c7', '#c8c8c8', '#c6c6c6', '#c6c6c6', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#a750ff', '#ffffff', '#ffffff', '#ffffff', '#a64dff', '#c1c1c1', '#c1c1c1', '#d2d2d2', '#ffffff', '#cdcdcd', '#ffffff', '#909090', '#c6c6c6', '#c6c6c6', '#898989', '#ffffff', '#c2c2c2', '#c2c2c2', '#c8c8c8', '#ffffff', '#ffffff', '#cacaca', '#c4c4c4', '#ffffff', '#ffffff', '#ffffff', '#a4a4a4', '#ffffff', '#666666', '#c4c4c4', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#a5a5a5', '#9a36ff', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c5c5c5', '#c6c6c6', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#a346ff', '#838383', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#a54aff', '#ffffff', '#ac5aff', '#ffffff', '#ffffff', '#ffffff', '#929292', '#ffffff', '#ffffff', '#838383', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#c7c7c7', '#cccccc', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#767676', '#c5c5c5', '#ffffff', '#ffffff', '#9a9a9a', '#959595', '#ffffff', '#c9c9c9', '#ffffff', '#ffffff', '#868686', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c7c7c7', '#c9c9c9', '#ffffff', '#ffffff', '#ffffff', '#ac59ff', '#ffffff', '#888888', '#949494', '#ffffff', '#cccccc', '#cacaca', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#646464', '#ffffff', '#9c9c9c', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c4c4c4', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#cccccc', '#ffffff', '#ffffff', '#c3c3c3', '#cccccc', '#ffffff', '#cbcbcb', '#ffffff', '#ffffff', '#ffffff', '#c8c8c8', '#ffffff', '#7f7f7f', '#d2d2d2', '#ffffff', '#ffffff', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#c4c4c4', '#ffffff', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#c6c6c6', '#646464', '#ffffff', '#ffffff', '#868686', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#7a7a7a', '#ffffff', '#ffffff', '#ffffff', '#c1c1c1', '#ffffff', '#ffffff', '#ffffff', '#5e5e5e', '#949494', '#ffffff', '#ffffff', '#797979', '#ffffff', '#c4c4c4', '#ffffff', '#a4a4a4', '#ffffff', '#ffffff', '#cccccc', '#ffffff', '#ffffff', '#838383', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#d0d0d0', '#ffffff', '#cccccc', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#686868', '#cacaca', '#c6c6c6', '#c4c4c4', '#cccccc', '#ffffff', '#ffffff', '#c6c6c6', '#ffffff', '#ffffff', '#c5c5c5', '#c1c1c1', '#ffffff', '#c3c3c3', '#ffffff', '#ffffff', '#ffffff', '#858585', '#8e8e8e', '#c1c1c1', '#c6c6c6', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#636363', '#a5a5a5', '#ffffff', '#ffffff', '#ffffff', '#858585', '#8f8f8f', '#ffffff', '#ffffff', '#878787', '#c5c5c5', '#ffffff', '#ffffff', '#cdcdcd', '#9e9e9e', '#c5c5c5', '#ffffff', '#ffffff', '#ffffff', '#a0a0a0', '#ffffff', '#ffffff', '#ffffff', '#c5c5c5', '#c3c3c3', '#ffffff', '#cbcbcb', '#ffffff', '#c9c9c9', '#ffffff', '#909090', '#cccccc', '#989898', '#c5c5c5', '#ffffff', '#ffffff', '#a4a4a4', '#ffffff', '#c7c7c7', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff'], ['4-hydroxytamoxifen, Dexamethasone, and Retinoic Acids Regulation of p27 Expression WP3879', 'ACE Inhibitor Pathway WP554', 'AGE/RAGE pathway WP2324', 'Sphingolipid pathway WP1422', 'ATM Signaling Network in Development and Disease  WP3878', 'ATM Signaling Pathway WP2516', 'ATR Signaling WP3875', 'Prion disease pathway WP3995', 'Sulfation Biotransformation Reaction WP692', 'EPO Receptor Signaling WP581', 'miRNA regulation of prostate cancer signaling pathways WP3981', 'The human immune response to tuberculosis WP4197', 'Alpha 6 Beta 4 signaling pathway WP244', 'GPCRs, Class B Secretin-like WP334', 'The alternative pathway of fetal androgen synthesis WP4524', 'Amplification and Expansion of Oncogenic Pathways as Metastatic Traits WP3678', 'miRNA Regulation of DNA Damage Response WP1530', 'Androgen receptor signaling pathway WP138', 'TCA Cycle and Deficiency of Pyruvate Dehydrogenase complex (PDHc) WP2453', 'Regulation of Apoptosis by Parathyroid Hormone-related Protein WP3872', 'ApoE and miR-146 in inflammation and atherosclerosis WP3926', 'Apoptosis-related network due to altered Notch3 in ovarian cancer WP2864', 'Apoptosis Modulation and Signaling WP1772', 'Apoptosis Modulation by HSP70 WP384', 'Apoptosis WP254', 'Arachidonate Epoxygenase / Epoxide Hydrolase WP678', 'Oxidative Damage WP3941', 'Aryl Hydrocarbon Receptor Pathway WP2873', 'Aryl Hydrocarbon Receptor WP2586', 'Arylamine metabolism WP694', 'Monoamine Transport WP727', 'BDNF-TrkB Signaling WP3676', 'BMP2-WNT4-FOXO1 Pathway in Human Primary Endometrial Stromal Cell Differentiation WP3876', 'BMP Signaling Pathway in Eyelid Development WP3927', 'B Cell Receptor Signaling Pathway WP23', 'Benzene metabolism WP3891', 'Benzo(a)pyrene metabolism WP696', 'PTF1A related regulatory pathway WP4147', 'Cell Cycle WP179', 'Cytosine methylation WP3585', 'Metastatic brain tumor WP2249', 'Peptide GPCRs WP24', 'Nicotine Activity on Dopaminergic Neurons WP1602', 'Breast cancer pathway WP4262', 'Alzheimers Disease WP2059', 'Caloric restriction and aging WP4191', 'miRNAs involvement in the immune response in sepsis WP4329', 'Bile Acids synthesis and enterohepatic circulation  WP4389', 'Canonical and Non-Canonical TGF-B signaling WP3874', 'Cardiac Hypertrophic Response WP2795', 'Biogenic Amine Synthesis WP550', '', 'Cell-type Dependent Selectivity of CCK2R Signaling WP3679', '', 'Cell Differentiation - Index WP2029', 'Cell Differentiation - Index expanded WP2023', 'Robo4 and VEGF Signaling Pathways Crosstalk WP3943', 'DDX1 as a regulatory component of the Drosha microprocessor WP2942', 'Chemokine signaling pathway WP3929', 'Cholesterol Biosynthesis Pathway WP197', 'Regulation of sister chromatid separation at the metaphase-anaphase transition WP4240', 'Mammary gland development pathway - Involution (Stage 4 of 4) WP2815', 'Metapathway biotransformation Phase I and II WP702', 'Classical pathway of steroidogenesis, including diseases WP4523', 'Cerebral Organic Acidurias, including diseases WP4519', 'Common Pathways Underlying Drug Addiction WP2636', 'Complement Activation WP545', 'Complement and Coagulation Cascades WP558', 'Toll-like Receptor Signaling WP3858', 'Structural Pathway of Interleukin 1 (IL-1) WP2637', 'Statin Pathway WP430', 'Target Of Rapamycin (TOR) Signaling WP1471', 'Transcription co-factors SKI and SKIL protein partners WP4533', 'miR-509-3p alteration of YAP1/ECM axis WP3967', 'Cytokines and Inflammatory Response WP530', 'Proteasome Degradation WP183', 'Phase I biotransformations, non P450 WP136', 'Gamma-Glutamyl Cycle for the biosynthesis and degradation of glutathione, including diseases WP4518', 'DNA Damage Response (only ATM dependent) WP710', 'DNA Damage Response WP707', 'DNA IR-Double Strand Breaks (DSBs) and cellular response via ATM WP3959', 'DNA IR-damage and cellular response via ATR WP4016', 'DNA Mismatch Repair WP531', 'DNA Replication WP466', 'Homologous recombination WP186', 'Synaptic Vesicle Pathway WP2267', 'Ectoderm Differentiation WP2858', 'Development of pulmonary dendritic cells and macrophage subsets WP3892', 'Microglia Pathogen Phagocytosis Pathway WP3937', 'Ferroptosis WP4313', 'Disorders of Folate Metabolism and Transport WP4259', 'Disorders of the Krebs cycle WP4236', 'Dopamine metabolism WP2436', 'Dopaminergic Neurogenesis WP2855', 'Type 2 papillary renal cell carcinoma WP4241', 'Dual hijack model of Vif in HIV infection WP3300', 'EBV LMP1 signaling WP262', 'NLR Proteins WP288', 'Simplified Interaction Map Between LOXL4 and Oxidative Stress Pathway WP3670', 'Aflatoxin B1 metabolism WP699', 'Epithelial to mesenchymal transition in colorectal cancer WP4239', 'ESC Pluripotency Pathways WP3931', 'EV release from cardiac cells and their functional effects WP3297', 'Cells and Molecules involved in local acute inflammatory response  WP4493', 'Development and heterogeneity of the ILC family WP3893', 'Effects of Nitric Oxide WP1995', 'Amino Acid metabolism WP3925', 'Electron Transport Chain (OXPHOS system in mitochondria) WP111', 'Endochondral Ossification WP474', 'Endoderm Differentiation WP2853', 'Endometrial cancer WP4155', 'Endothelin Pathways WP2197', 'Energy Metabolism WP1541', \"ERK Pathway in Huntington's Disease WP3853\", 'Thermogenesis WP4321', 'Tumor suppressor activity of SMARCB1 WP4204', 'GPR40 Pathway WP3958', 'Estrogen signaling pathway WP712', 'Ethanol effects on histone modifications WP3996', 'Ethanol metabolism resulting in production of ROS by CYP2E1 WP4269', 'Sulindac Metabolic Pathway WP2542', 'Regulation of Wnt/B-catenin Signaling by Small Molecule Compounds WP3664', 'Extracellular vesicle-mediated signaling in recipient cells WP2870', 'Splicing factor NOVA regulated synaptic proteins WP4148', 'Toll-like Receptor Signaling Pathway WP75', 'Factors and pathways affecting insulin-like growth factor (IGF1)-Akt signaling WP3850', 'Farnesoid X Receptor  Pathway WP2879', 'Blood Clotting Cascade WP272', 'Fatty Acid Beta Oxidation WP143', 'Fatty Acid Biosynthesis WP357', 'Folate Metabolism WP176', 'Preimplantation Embryo WP3527', 'mir-124 predicted interactions with cell cycle and differentiation  WP3595', 'Angiogenesis WP1539', 'Valproic acid pathway WP3871', 'Focal Adhesion WP306', 'Folate-Alcohol and Cancer Pathway Hypotheses WP1589', 'NRF2 pathway WP2884', 'Follicle Stimulating Hormone (FSH) signaling pathway WP2035', 'G1 to S cell cycle control WP45', 'TP53 Network WP1742', 'GABA receptor Signaling WP4159', 'GPCRs, Class A Rhodopsin-like WP455', 'Zinc homeostasis WP3529', 'GPCRs, Class C Metabotropic glutamate, pheromone WP501', 'GPCRs, Other WP117', 'Monoamine GPCRs WP58', 'Prolactin Signaling Pathway WP2037', 'Angiopoietin Like Protein 8 Regulatory Pathway WP3915', 'Ganglio Sphingolipid Metabolism WP1423', 'Gastric Cancer Network 1 WP2361', 'Gastric Cancer Network 2 WP2363', 'Gastric acid production WP2596', 'NAD metabolism, sirtuins and aging WP3630', 'Bone Morphogenic Protein (BMP) Signalling and Regulation WP1425', 'Genes targeted by miRNAs in adipocytes WP1992', 'Genotoxicity pathway WP4286', 'Glial Cell Differentiation WP2276', 'One Carbon Metabolism WP241', 'Secretion of Hydrochloric Acid in Parietal Cells WP2597', 'TYROBP Causal Network WP3945', 'Glutathione metabolism WP100', 'NO/cGMP/PKG mediated Neuroprotection WP4008', 'Glycogen Synthesis and Degradation WP500', 'Drug Induction of Bile Acid Pathway WP2289', 'Glycosylation and related congenital defects. WP4521', 'Acetylcholine Synthesis WP528', 'HIF1A and PPARG regulation of glycolysis WP2456', 'ncRNAs involved in STAT3 signaling in hepatocellular carcinoma WP4337', 'Hedgehog Signaling Pathway WP4249', 'Hedgehog Signaling Pathway WP47', 'Metabolism of Spingolipids in ER and Golgi apparatus WP4142', 'Hematopoietic Stem Cell Differentiation WP2849', 'Estrogen metabolism WP697', 'Adipogenesis WP236', 'Leptin and adiponectin WP3934', 'Hereditary leiomyomatosis and renal cell carcinoma pathway WP4206', 'miRs in Muscle Cell Differentiation WP2012', 'Hippo-Merlin Signaling Dysregulation WP4541', 'Hippo-Yap signaling pathway WP4537', 'Histone Modifications WP2369', 'Metabolic pathway of LDL, HDL and TG, including diseases WP4522', 'Human Complement System WP2806', 'Human Thyroid Stimulating Hormone (TSH) signaling pathway WP2032', 'Hypertrophy Model WP516', 'Hypothesized Pathways in Pathogenesis of Cardiovascular Disease WP3668', 'Nanoparticle-mediated activation of receptor signaling WP2643', '', 'IL-1 signaling pathway WP195', 'IL-2 Signaling Pathway WP49', 'IL-3 Signaling Pathway WP286', 'IL-4 Signaling Pathway WP395', 'IL-5 Signaling Pathway WP127', 'IL-6 signaling pathway WP364', 'IL-7 Signaling Pathway WP205', 'IL-9 Signaling Pathway WP22', 'IL-10 Anti-inflammatory Signaling Pathway  WP4495', 'Type III interferon signaling WP2113', 'IL17 signaling pathway WP2112', 'Focal Adhesion-PI3K-Akt-mTOR-signaling pathway WP3932', 'Inflammatory Response Pathway WP453', 'Inhibition of exosome biogenesis and secretion by Manumycin A in CRPC cells WP4301', 'Initiation of transcription and translation elongation at the HIV-1 LTR WP3414', 'Insulin Signaling WP481', 'Insulin signalling in human adipocytes (diabetic condition) WP3635', 'Insulin signalling in human adipocytes (normal condition) WP3634', 'Integrated Breast Cancer Pathway WP1984', 'Integrated Cancer Pathway WP1971', 'Integrin-mediated Cell Adhesion WP185', 'Interactome of polycomb repressive complex 2 (PRC2)  WP2916', 'Interferon type I signaling pathways WP585', 'Bladder Cancer WP2828', 'Interleukin-11 Signaling Pathway WP2332', 'Intraflagellar transport proteins binding to dynein WP4532', 'Hfe effect on hepcidin production WP3924', 'Iron metabolism in placenta WP2007', 'Kennedy pathway from Sphingolipids WP3933', 'Kit receptor signaling pathway WP304', 'Leptin Insulin Overlap WP3935', 'Hepatitis C and Hepatocellular Carcinoma WP3646', 'Leptin signaling pathway WP2034', 'Lipid Metabolism Pathway WP3965', 'Liver X Receptor Pathway WP2874', 'Cysteine and methionine catabolism WP4504', 'Differentiation Pathway WP2848', 'Calcium Regulation in the Cardiac Cell WP536', 'MAPK Cascade WP422', 'MAPK Signaling Pathway WP382', 'MTHFR deficiency WP4288', 'Fas Ligand (FasL) pathway and Stress induction of Heat Shock Proteins (HSP) regulation WP314', 'MET in type 1 papillary renal cell carcinoma WP4205', 'MFAP5-mediated ovarian cancer cell motility and invasiveness WP3301', 'Overview of nanoparticle effects WP3287', 'Macrophage markers WP4146', 'Nonalcoholic fatty liver disease WP4396', 'Ciliary landscape WP4352', 'Mammary gland development pathway - Pregnancy and lactation (Stage 3 of 4) WP2817', 'Mammary gland development pathway - Puberty (Stage 2 of 4) WP2814', 'Matrix Metalloproteinases WP129', 'Mechanoregulation and pathology of YAP/TAZ via Hippo and non-Hippo mechanisms WP4534', 'Nucleotide Metabolism WP404', 'Prader-Willi and Angelman Syndrome WP3998', 'Oxidative Stress WP408', 'Metabolic reprogramming in colon cancer WP4290', 'Heart Development WP1591', 'Phytochemical activity on NRF2 transcriptional activation WP3', 'MECP2 and Associated Rett Syndrome WP3584', 'Methionine De Novo and Salvage Pathway WP3580', 'Methionine metabolism leading to Sulphur Amino Acids and related disorders WP4292', 'Methylation Pathways WP704', 'Mevalonate pathway WP3963', 'MicroRNA for Targeting Cancer Growth and Vascularization in Glioblastoma WP3593', 'MicroRNAs in cardiomyocyte hypertrophy WP1544', 'Role of Osx and miRNAs in tooth development WP3971', 'Mitochondrial Gene Expression WP391', 'Catalytic cycle of mammalian Flavin-containing MonoOxygenases (FMOs) WP688', 'Mitochondrial complex I assembly model OXPHOS system WP4324', '', 'Hematopoietic Stem Cell Gene Regulation by GABP alpha/beta Complex WP3657', 'Oxysterols derived from cholesterol WP4545', 'Oxidative phosphorylation WP623', 'NAD+ biosynthetic pathways WP3645', 'NAD+ metabolism WP3644', 'NAD Biosynthesis II (from tryptophan) WP2485', 'Fatty Acid Omega Oxidation WP206', 'Tamoxifen metabolism WP691', 'Glycerophospholipid Biosynthetic Pathway WP2533', 'NOTCH1 regulation of human endothelial cell calcification WP3413', 'Synaptic signaling pathways associated with autism spectrum disorder WP4539', 'Allograft Rejection WP2328', 'Nanomaterial induced apoptosis WP2507', 'Nucleotide GPCRs WP80', 'Nanoparticle triggered autophagic cell death WP2509', 'Nanoparticle triggered regulated necrosis WP2513', 'Neovascularisation processes WP4331', 'Neural Crest Differentiation WP2064', '', 'Nicotine Metabolism WP1600', 'Supression of HMGB1 mediated inflammation by THBD WP4479', 'Non-homologous end joining WP438', 'PI3K/AKT/mTOR - VitD3 Signalling WP4141', 'Mammary gland development pathway - Embryonic development (Stage 1 of 4) WP2813', 'Notch Signaling Pathway WP61', 'Notch Signaling WP268', 'Nuclear Receptors Meta-Pathway WP2882', 'Nuclear Receptors WP170', 'Nuclear Receptors in Lipid Metabolism and Toxicity WP299', 'Nucleotide-binding Oligomerization Domain (NOD) pathway WP1433', 'LncRNA involvement in canonical Wnt signaling and colorectal cancer WP4258', 'Primary Focal Segmental Glomerulosclerosis FSGS WP2572', 'The effect of progerin on the involved genes in Hutchinson-Gilford Progeria Syndrome WP4320', 'Oncostatin M Signaling Pathway WP2374', 'Hypothetical Craniofacial Development Pathway WP3655', 'Ebola Virus Pathway on Host WP4217', 'Mitochondrial LC-Fatty Acid Beta-Oxidation WP368', 'Osteoclast Signaling WP12', 'Osteopontin Signaling WP1434', 'Ovarian Infertility Genes WP34', 'MAPK  and NFkB Signalling Pathways Inhibited by Yersinia YopJ WP3849', 'Rett syndrome causing genes WP4312', 'Eicosanoid Synthesis WP167', 'Degradation pathway of sphingolipids, including diseases WP4153', 'Translation Factors WP107', 'Association Between Physico-Chemical Features and Toxicity Associated Pathways WP3680', 'PDGFR-beta pathway WP3972', 'PDGF Pathway WP2526', 'PI3K-AKT-mTOR signaling pathway and therapeutic opportunities WP3844', 'PI3K-Akt Signaling Pathway WP4172', 'Non-small cell lung cancer WP4255', 'PPAR Alpha Pathway WP2878', 'PPAR signaling pathway WP3942', 'H19 action Rb-E2F1 signaling and CDK-Beta-catenin activity WP3969', 'Pancreatic adenocarcinoma pathway WP4263', 'Parkin-Ubiquitin Proteasomal System pathway WP2359', 'Parkinsons Disease Pathway WP2371', 'Pathogenic Escherichia coli infection WP2272', 'Eukaryotic Transcription Initiation WP405', 'Osteoblast Signaling WP322', 'Pathways in clear cell renal cell carcinoma WP4018', 'Pentose Phosphate Metabolism WP134', 'Genes related to primary cilium development (based on CRISPR) WP4536', 'Interleukin-1 Induced Activation of NF-kappa-B WP3656', 'Molybdenum cofactor (Moco) biosynthesis WP4507', '', 'Photodynamic therapy-induced HIF-1 survival signaling WP3614', 'Thiamine metabolic pathways WP4297', 'Thyroxine (Thyroid Hormone) Production WP1981', 'Photodynamic therapy-induced unfolded protein response WP3613', 'Somatroph axis (GH) and its relationship to dietary restriction and aging WP4186', 'TCA Cycle (aka Krebs or citric acid cycle) WP78', 'Platelet-mediated interactions with vascular and circulating cells WP4462', 'Deregulation of Rab and Rab Effector Genes in Bladder Cancer WP2291', 'Pregnane X Receptor pathway WP2876', 'Irinotecan Pathway WP229', 'Corticotropin-releasing hormone signaling pathway WP2355', 'Canonical  and Non-canonical Notch signaling WP3845', 'T-Cell Receptor and Co-stimulatory Signaling WP2583', 'Extracellular vesicles in the crosstalk of cardiac cells WP4300', 'Cytoplasmic Ribosomal Proteins WP477', 'Purine metabolism WP4224', 'Pyrimidine metabolism WP4022', 'Pyrimidine metabolism and related diseases WP4225', 'Melatonin metabolism and effects WP3298', 'RANKL/RANK (Receptor activator of NFKB (ligand)) Signaling Pathway WP2018', 'RIG-I-like Receptor Signaling WP3865', 'RalA downstream regulated genes WP2290', 'Ras Signaling WP4223', 'Regulation of Actin Cytoskeleton WP51', 'Codeine and Morphine Metabolism WP1604', 'Photodynamic therapy-induced NFE2L2 (NRF2) survival signaling WP3612', 'G13 Signaling Pathway WP524', 'Chromosomal and microsatellite instability in colorectal cancer  WP4216', 'Regulation of toll-like receptor signaling pathway WP1449', 'Resistin as a regulator of inflammation WP4481', 'Striated Muscle Contraction Pathway WP383', 'Oxidation by Cytochrome P450 WP43', 'Fibrin Complement Receptor 3 Signaling Pathway WP4136', 'Role Altered Glycolysation of MUC1 in Tumour Microenvironment WP4480', 'Glucuronidation WP698', 'Estrogen Receptor Pathway WP2881', 'SREBF and miR33 in cholesterol and lipid homeostasis WP2011', 'SRF and miRs in Smooth Muscle Differentiation and Proliferation WP1991', 'Glucocorticoid and Mineralcorticoid Metabolism WP237', 'Selective expression of chemokine receptors during T-cell polarization WP4494', 'Selenium Metabolism and Selenoproteins WP28', 'Selenium Micronutrient Network WP15', 'Senescence and Autophagy in Cancer WP615', 'Serotonin Receptor 2 and ELK-SRF/GATA4 signaling WP732', 'Serotonin Receptor 4/6/7 and NR3C Signaling WP734', 'Serotonin Transporter Activity WP1455', 'Serotonin and anxiety-related events WP3944', 'Serotonin and anxiety WP3947', 'Simplified Depiction of MYD88 Distinct Input-Output Pathway WP3877', 'Signal transduction through IL1R WP4496', 'Signaling Pathways in Glioblastoma WP2261', 'Signaling of Hepatocyte Growth Factor Receptor WP313', 'Signal Transduction of S1P Receptor WP26', 'EGF/EGFR Signaling Pathway WP437', 'Sleep regulation WP3591', 'Small Ligand GPCRs WP247', 'Gene regulatory network modelling somitogenesis  WP2854', 'Steroid Biosynthesis WP496', 'Spinal Cord Injury WP2431', 'Prostaglandin Synthesis and Regulation WP98', 'Copper homeostasis WP3286', 'AMP-activated Protein Kinase (AMPK) Signaling WP1403', 'Sterol Regulatory Element-Binding Proteins (SREBP) signalling WP1982', 'Retinoblastoma Gene in Cancer WP2446', 'Constitutive Androstane Receptor Pathway WP2875', 'Sudden Infant Death Syndrome (SIDS) Susceptibility Pathways WP706', 'SCFA and skeletal muscle substrate metabolism WP4030', 'Cannabinoid receptor signaling WP3869', 'Non-genomic actions of 1,25 dihydroxyvitamin D3 WP4341', 'Mesodermal Commitment Pathway WP2857', 'NRF2-ARE regulation WP4357', 'Synthesis and Degradation of Ketone Bodies WP311', 'G Protein Signaling Pathways WP35', 'T-Cell antigen Receptor (TCR)  Signaling Pathway WP69', 'T-Cell antigen Receptor (TCR) pathway during Staphylococcus aureus infection WP3863', 'Circadian rhythm related genes WP3594', '', 'Fluoropyrimidine Activity WP1601', 'TFs Regulate miRNAs related to cardiac hypertrophy WP1559', 'TGF-B Signaling in Thyroid Cells for Epithelial-Mesenchymal Transition WP3859', 'TGF-beta Receptor Signaling WP560', 'TGF-beta Signaling Pathway WP366', 'TLR4 Signaling and Tolerance WP3851', 'TNF alpha Signaling Pathway WP231', 'TNF related weak inducer of apoptosis (TWEAK) Signaling Pathway WP2036', 'Oligodendrocyte Specification and differentiation(including remyelination), leading to Myelin Components for CNS WP4304', 'Globo Sphingolipid Metabolism WP1424', 'EDA Signalling in Hair Follicle Development WP3930', 'Cori Cycle WP1946', 'Pathways Regulating Hippo Signaling WP4540', 'Arrhythmogenic Right Ventricular Cardiomyopathy WP2118', 'Vitamin A and Carotenoid Metabolism WP716', 'Physiological and Pathological Hypertrophy  of the Heart WP1528', 'ErbB Signaling Pathway WP673', 'Photodynamic therapy-induced NF-kB survival signaling WP3617', 'Thymic Stromal LymphoPoietin (TSLP) Signaling Pathway WP2203', 'Regulation of Microtubule Cytoskeleton WP2038', 'FTO Obesity Variant Mechanism WP3407', 'Composition of Lipid Particles WP3601', 'Trans-sulfuration and one carbon metabolism WP2525', 'Trans-sulfuration pathway WP2333', 'RAC1/PAK1/p38/MMP2 Pathway WP3303', 'Transcription factor regulation in adipogenesis WP3599', 'Transcriptional cascade regulating adipogenesis WP4211', 'Tgif disruption of Shh signaling WP3674', 'Exercise-induced Circadian Regulation WP410', 'Tryptophan catabolism leading to NAD+ production WP4210', 'Tryptophan metabolism WP465', 'Heme Biosynthesis WP561', 'Glycolysis and Gluconeogenesis WP534', 'IL1 and megakaryocytes in obesity WP2865', 'Type II diabetes mellitus WP1584', 'miRNAs involved in DNA damage response WP1545', 'Urea cycle and metabolism of amino groups WP497', 'VEGFA-VEGFR2 Signaling Pathway WP3888', 'Myometrial Relaxation and Contraction Pathways WP289', 'miRNA targets in ECM and membrane receptors WP2911', 'Triacylglyceride Synthesis WP325', 'Vitamin B6-dependent and responsive disorders WP4228', 'Vitamin B12 Disorders WP4271', 'Vitamin B12 Metabolism WP1533', 'Vitamin D Metabolism WP1531', 'Vitamin D Receptor Pathway WP2877', 'Vitamin D in inflammatory diseases WP4482', 'White fat cell differentiation WP4149', 'Wnt/beta-catenin Signaling Pathway in Leukemia WP3658', 'Wnt Signaling Pathway WP363', 'Wnt Signaling Pathway and Pluripotency WP399', 'Wnt Signaling WP428', 'Wnt Signaling in Kidney Disease WP4150', 'Lung fibrosis WP3624', 'exRNA mechanism of action and biogenesis WP2805', 'let-7 inhibition of ES cell reprogramming WP3299', 'mRNA Processing WP411', 'LncRNA-mediated mechanisms of therapeutic resistance WP3672', 'miR-517 relationship with ARCN1 and USP1 WP3596', 'miRNA Biogenesis WP2338', 'Amyotrophic lateral sclerosis (ALS) WP2447', 'miRNA regulation of p53 pathway in prostate cancer WP3982', 'Alanine and aspartate metabolism WP106', 'Viral Acute Myocarditis WP4298', 'Type II interferon signaling (IFNG) WP619', 'Pathways Affected in Adenoid Cystic Carcinoma WP3651', 'Differentiation of white and brown adipocyte    WP2895', 'One carbon metabolism and related pathways WP3940', 'Hair Follicle Development: Cytodifferentiation (Part 3 of 3) WP2840', 'ncRNAs involved in Wnt signaling in hepatocellular carcinoma WP4336', 'p38 MAPK Signaling Pathway WP400', 'TCA Cycle Nutrient Utilization and Invasiveness of Ovarian Cancer WP2868', '', 'Brain-Derived Neurotrophic Factor (BDNF) signaling pathway WP2380', 'ID signaling pathway WP53', 'Cardiac Progenitor Differentiation WP2406', '', 'Photodynamic therapy-induced AP-1 survival signaling. WP3611', '', '', '', 'Phosphodiesterases in neuronal function WP4222', 'Imatinib and Chronic Myeloid Leukemia WP3640'], [0.0, 0.0, 0.0, 0.00249, 0.0, 0.00241, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00251, 0.0, 0.0, 0.00259, 0.0, 0.0, 0.00215, 0.00512, 0.0, 0.0, 0.00234, 0.00215, 0.0, 0.00218, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00244, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00253, 0.00378, 0.0, 0.0, 0.0, 0.00258, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00262, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00717, 0.00256, 0.00262, 0.0024, 0.00252, 0.00231, 0.0, 0.00236, 0.0, 0.00243, 0.0, 0.0, 0.0, 0.00458, 0.0, 0.0, 0.00412, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00195, 0.0, 0.0, 0.00241, 0.0, 0.00262, 0.00258, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00562, 0.00611, 0.0, 0.0, 0.0, 0.0, 0.01073, 0.00628, 0.00228, 0.00194, 0.0, 0.00245, 0.00474, 0.00257, 0.00621, 0.00246, 0.0, 0.00251, 0.00246, 0.0, 0.0, 0.0, 0.0, 0.00719, 0.0, 0.00246, 0.0, 0.0, 0.00244, 0.0, 0.00227, 0.0, 0.0, 0.00251, 0.0, 0.00524, 0.0026, 0.0, 0.00498, 0.0, 0.0, 0.00246, 0.0, 0.00243, 0.0, 0.00215, 0.0, 0.00222, 0.00394, 0.00258, 0.00248, 0.0, 0.0, 0.00259, 0.0, 0.0, 0.00685, 0.0, 0.00494, 0.0, 0.0, 0.0, 0.0, 0.00482, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00477, 0.00256, 0.0, 0.00233, 0.0, 0.00594, 0.0026, 0.0, 0.00248, 0.00606, 0.0, 0.00226, 0.0, 0.00424, 0.00227, 0.0, 0.0, 0.00248, 0.0, 0.0, 0.0024, 0.00236, 0.00233, 0.00241, 0.00239, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00737, 0.0, 0.0, 0.0, 0.00748, 0.00261, 0.00261, 0.0019, 0.0, 0.0021, 0.0, 0.00466, 0.00241, 0.00239, 0.00498, 0.0, 0.00258, 0.00258, 0.0023, 0.0, 0.0, 0.00222, 0.00248, 0.0, 0.0, 0.0, 0.00382, 0.0, 0.00644, 0.0025, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00377, 0.00846, 0.00246, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00245, 0.0024, 0.0, 0.0, 0.0, 0.00252, 0.00777, 0.00521, 0.0, 0.0, 0.0, 0.0, 0.00761, 0.0, 0.00696, 0.0, 0.0, 0.0, 0.0046, 0.0, 0.0, 0.00522, 0.0, 0.0, 0.00253, 0.0, 0.00235, 0.00216, 0.0, 0.0, 0.0, 0.0, 0.00253, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00576, 0.00242, 0.0, 0.0, 0.00426, 0.00447, 0.0, 0.00227, 0.0, 0.0, 0.0051, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00234, 0.00227, 0.0, 0.0, 0.0, 0.00699, 0.0, 0.00499, 0.00452, 0.0, 0.00216, 0.00225, 0.00242, 0.0, 0.0, 0.0, 0.00652, 0.0, 0.00418, 0.0, 0.00262, 0.0, 0.0, 0.0, 0.0, 0.00249, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00214, 0.0, 0.0, 0.00254, 0.00216, 0.0, 0.00217, 0.0, 0.0, 0.0, 0.0023, 0.0, 0.00537, 0.0019, 0.0, 0.0, 0.00242, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00249, 0.0, 0.0, 0.0, 0.00261, 0.0, 0.00238, 0.00651, 0.0, 0.0, 0.00508, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00559, 0.0, 0.0, 0.0, 0.0026, 0.0, 0.0, 0.0, 0.00676, 0.0045, 0.0, 0.0, 0.00563, 0.0, 0.00248, 0.0, 0.00383, 0.0, 0.0, 0.00214, 0.0, 0.0, 0.00521, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00197, 0.0, 0.00214, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00634, 0.00223, 0.00239, 0.0025, 0.00215, 0.0, 0.0, 0.00238, 0.0, 0.0, 0.00246, 0.0026, 0.0, 0.00252, 0.0, 0.0, 0.0, 0.00512, 0.00475, 0.0026, 0.00238, 0.0, 0.0, 0.0, 0.0, 0.00655, 0.00377, 0.0, 0.0, 0.0, 0.00515, 0.00471, 0.0, 0.0, 0.00504, 0.00246, 0.0, 0.0, 0.0021, 0.00408, 0.00243, 0.0, 0.0, 0.0, 0.00399, 0.0, 0.0, 0.0, 0.00242, 0.00251, 0.0, 0.00218, 0.0, 0.00227, 0.0, 0.00468, 0.00216, 0.00434, 0.00244, 0.0, 0.0, 0.00385, 0.0, 0.00234, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])\n",
       "})\n",
       "</script>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The color scaling factor for this canvas is: 0.01073\n",
      "Gene sets with a similarity index less than 0.00696 are not colored\n"
     ]
    }
   ],
   "source": [
    "if len(all_libraries) == 1:\n",
    "    anneal_list, x_dimension, y_dimension = library_processing(0)\n",
    "    anneal_list = annealing(anneal_list, int(annealing_steps), find_fitness(anneal_list, x_dimension, y_dimension)[0], x_dimension, y_dimension)\n",
    "    color_list, scaling_factor, cut_off_value = get_color(anneal_list, 2.0, scaling_factor, x_dimension, y_dimension)\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "    Canvas(x_dimension, y_dimension, color_list, list(unzipped_anneal_list[0]), list(unzipped_anneal_list[2]))\n",
    "    print(\"The color scaling factor for this canvas is: \" + str(scaling_factor))\n",
    "    print(\"Gene sets with a similarity index less than \" + str(cut_off_value) + \" are not colored\")\n",
    "\n",
    "elif len(all_libraries) == 2:\n",
    "    # first library\n",
    "    anneal_list_1, x_dimension_1, y_dimension_1 = library_processing(0)\n",
    "    anneal_list_1 = annealing(anneal_list_1, int(annealing_steps), find_fitness(anneal_list_1, x_dimension_1, y_dimension_1)[0], x_dimension_1, y_dimension_1)\n",
    "    color_list_1, scaling_factor_1, cut_off_value_1 = get_color(anneal_list_1, 2.0, scaling_factor, x_dimension_1, y_dimension_1)\n",
    "\n",
    "    # second library\n",
    "    anneal_list_2, x_dimension_2, y_dimension_2 = library_processing(1)\n",
    "\n",
    "    anneal_list_2 = annealing(anneal_list_2, int(annealing_steps), find_fitness(anneal_list_2, x_dimension_2, y_dimension_2)[0], x_dimension_2, y_dimension_2)\n",
    "    color_list_2, scaling_factor_2, cut_off_value_2 = get_color(anneal_list_2, 2.0, scaling_factor, x_dimension_2, y_dimension_2)\n",
    "\n",
    "    if scaling_factor_1 > scaling_factor_2:\n",
    "        scaling_factor = scaling_factor_1\n",
    "        color_list_2, scaling_factor_2, cut_off_value_2 = get_color(anneal_list_2, 2.0, scaling_factor, x_dimension_2, y_dimension_2)\n",
    "    else:\n",
    "        scaling_factor = scaling_factor_2\n",
    "        color_list_1, scaling_factor_1, cut_off_value_1 = get_color(anneal_list_1, 2.0, scaling_factor, x_dimension_1, y_dimension_1)\n",
    "\n",
    "    unzipped_anneal_list_1 = unzip_list(anneal_list_1)\n",
    "    unzipped_anneal_list_2 = unzip_list(anneal_list_2)\n",
    "\n",
    "    Canvas(x_dimension_1, y_dimension_1, color_list_1, list(unzipped_anneal_list_1[0]), list(unzipped_anneal_list_1[2]))\n",
    "    Canvas(x_dimension_2, y_dimension_2, color_list_2, list(unzipped_anneal_list_2[0]), list(unzipped_anneal_list_2[2]))\n",
    "\n",
    "    print(\"The color scaling factor for these canvases are: \" + str(scaling_factor))\n",
    "    print(\"For the first canvas, gene sets with a similarity index less than \" + str(cut_off_value_1) + \" are not colored\")\n",
    "    print(\"For the second canvas, gene sets with a similarity index less than \" + str(cut_off_value_2) + \" are not colored\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 64-bit ('py383': conda)",
   "language": "python",
   "name": "python38364bitpy383conda2309aa976eb546f3b7675a3d2adb70e0"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
